*** src/nsterm.h	29 Jan 2009 10:35:55 -0000	1.23
--- src/nsterm.h	20 Feb 2009 21:37:06 -0000
***************
*** 403,409 ****
  #define KEY_NS_NEW_FRAME               ((1<<28)|(0<<16)|12)
  #define KEY_NS_TOGGLE_TOOLBAR          ((1<<28)|(0<<16)|13)
  #define KEY_NS_INFO_PREFS              ((1<<28)|(0<<16)|14)
! 
  /* could use list to store these, but rest of emacs has a big infrastructure
     for managing a table of bitmap "records" */
  struct ns_bitmap_record
--- 404,413 ----
  #define KEY_NS_NEW_FRAME               ((1<<28)|(0<<16)|12)
  #define KEY_NS_TOGGLE_TOOLBAR          ((1<<28)|(0<<16)|13)
  #define KEY_NS_INFO_PREFS              ((1<<28)|(0<<16)|14)
! #define KEY_NS_SHOW_PREFS              ((1<<28)|(0<<16)|129)
! #define KEY_NS_ABOUT                   ((1<<28)|(0<<16)|130)
! #define KEY_NS_CHECK_FOR_UPDATES       ((1<<28)|(0<<16)|131)
!   
  /* could use list to store these, but rest of emacs has a big infrastructure
     for managing a table of bitmap "records" */
  struct ns_bitmap_record
*** ../emacs.temp/src/nsterm.m	2009-01-02 13:59:45.000000000 +1300
--- src/nsterm.m	2009-01-02 13:59:45.000000000 +1300
***************
*** 3938,3981 ****
      /* set up the application menu */
      svcsMenu = [[EmacsMenu alloc] initWithTitle: @"Services"];
      [svcsMenu setAutoenablesItems: NO];
!     appMenu = [[EmacsMenu alloc] initWithTitle: @"Emacs"];
      [appMenu setAutoenablesItems: NO];
      mainMenu = [[EmacsMenu alloc] initWithTitle: @""];
      dockMenu = [[EmacsMenu alloc] initWithTitle: @""];
  
!     [appMenu insertItemWithTitle: @"About Emacs"
!                           action: @selector (orderFrontStandardAboutPanel:)
                     keyEquivalent: @""
                           atIndex: 0];
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 1];
      [appMenu insertItemWithTitle: @"Preferences..."
                            action: @selector (showPreferencesWindow:)
                     keyEquivalent: @","
!                          atIndex: 2];
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 3];
      item = [appMenu insertItemWithTitle: @"Services"
                                   action: @selector (menuDown:)
                            keyEquivalent: @""
!                                 atIndex: 4];
      [appMenu setSubmenu: svcsMenu forItem: item];
  /*    [svcsMenu setSupercell: item]; */
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 5];
!     [appMenu insertItemWithTitle: @"Hide Emacs"
                            action: @selector (hide:)
                     keyEquivalent: @"h"
!                          atIndex: 6];
      item =  [appMenu insertItemWithTitle: @"Hide Others"
                            action: @selector (hideOtherApplications:)
                     keyEquivalent: @"h"
!                          atIndex: 7];
      [item setKeyEquivalentModifierMask: NSCommandKeyMask | NSAlternateKeyMask];
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 8];
!     [appMenu insertItemWithTitle: @"Quit Emacs"
                            action: @selector (terminate:)
                     keyEquivalent: @"q"
!                          atIndex: 9];
  
!     item = [mainMenu insertItemWithTitle: @"Emacs"
                                    action: @selector (menuDown:)
                             keyEquivalent: @""
                                   atIndex: 0];
--- 3967,4014 ----
      /* set up the application menu */
      svcsMenu = [[EmacsMenu alloc] initWithTitle: @"Services"];
      [svcsMenu setAutoenablesItems: NO];
!     appMenu = [[EmacsMenu alloc] initWithTitle: @"Aquamacs"];
      [appMenu setAutoenablesItems: NO];
      mainMenu = [[EmacsMenu alloc] initWithTitle: @""];
      dockMenu = [[EmacsMenu alloc] initWithTitle: @""];
  
!     [appMenu insertItemWithTitle: @"About Aquamacs Emacs"
!                           action: @selector (showAbout:)
                     keyEquivalent: @""
                           atIndex: 0];
!     [appMenu insertItemWithTitle: @"Check for Updates..."
!                           action: @selector (checkForUpdates:)
!                    keyEquivalent: @""
!                          atIndex: 1];
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 2];
      [appMenu insertItemWithTitle: @"Preferences..."
                            action: @selector (showPreferencesWindow:)
                     keyEquivalent: @","
!                          atIndex: 3];
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 4];
      item = [appMenu insertItemWithTitle: @"Services"
                                   action: @selector (menuDown:)
                            keyEquivalent: @""
!                                 atIndex: 5];
      [appMenu setSubmenu: svcsMenu forItem: item];
  /*    [svcsMenu setSupercell: item]; */
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 6];
!     [appMenu insertItemWithTitle: @"Hide Aquamacs Emacs"
                            action: @selector (hide:)
                     keyEquivalent: @"h"
!                          atIndex: 7];
      item =  [appMenu insertItemWithTitle: @"Hide Others"
                            action: @selector (hideOtherApplications:)
                     keyEquivalent: @"h"
!                          atIndex: 8];
      [item setKeyEquivalentModifierMask: NSCommandKeyMask | NSAlternateKeyMask];
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 9];
!     [appMenu insertItemWithTitle: @"Quit Aquamacs Emacs"
                            action: @selector (terminate:)
                     keyEquivalent: @"q"
!                          atIndex: 10];
  
!     item = [mainMenu insertItemWithTitle: @"Aquamacs Emacs"
                                    action: @selector (menuDown:)
                             keyEquivalent: @""
                                   atIndex: 0];
***************
*** 4103,4113 ****
  }
  
  
  - (void)showPreferencesWindow: (id)sender
  {
!   if (prefsController == nil)
!     prefsController = [[EmacsPrefsController alloc] init];
!   [prefsController showForFrame: SELECTED_FRAME ()];
  }
  
  
--- 4136,4179 ----
  }
  
  
+ - (void)checkForUpdates: (id)sender
+ {
+   struct frame *emacsframe = SELECTED_FRAME ();
+   NSEvent *theEvent = [NSApp currentEvent];
+ 
+   if (!emacs_event)
+     return;
+   emacs_event->kind = NON_ASCII_KEYSTROKE_EVENT;
+   emacs_event->code = KEY_NS_CHECK_FOR_UPDATES;
+   emacs_event->modifiers = 0;
+   EV_TRAILER (theEvent);
+ }
+ 
+ - (void)showAbout: (id)sender
+ {
+   struct frame *emacsframe = SELECTED_FRAME ();
+   NSEvent *theEvent = [NSApp currentEvent];
+ 
+   if (!emacs_event)
+     return;
+   emacs_event->kind = NON_ASCII_KEYSTROKE_EVENT;
+   emacs_event->code = KEY_NS_ABOUT;
+   emacs_event->modifiers = 0;
+   EV_TRAILER (theEvent);
+ }
+ 
+ 
  - (void)showPreferencesWindow: (id)sender
  {
!   struct frame *emacsframe = SELECTED_FRAME ();
!   NSEvent *theEvent = [NSApp currentEvent];
! 
!   if (!emacs_event)
!     return;
!   emacs_event->kind = NON_ASCII_KEYSTROKE_EVENT;
!   emacs_event->code = KEY_NS_SHOW_PREFS;
!   emacs_event->modifiers = 0;
!   EV_TRAILER (theEvent);
  }
  
  
*** lisp/nsmenu.m.~1.20.~	Tue Feb 24 23:05:47 2009
--- lisp/nsmenu.m	Mon Mar  2 12:10:23 2009
***************
*** 156,162 ****
  
    if (menu == nil)
      {
!       menu = [[EmacsMenu alloc] initWithTitle: @"Emacs"];
        needsSet = YES;
      }
    else
--- 156,162 ----
  
    if (menu == nil)
      {
!       menu = [[EmacsMenu alloc] initWithTitle: @"Aquamacs/Cocoa"];
        needsSet = YES;
      }
    else
***************
*** 174,180 ****
    /* widget_value is a straightforward object translation of emacs's
       Byzantine lisp menu structures */
    wv = xmalloc_widget_value ();
!   wv->name = "Emacs";
    wv->value = 0;
    wv->enabled = 1;
    wv->button_type = BUTTON_TYPE_NONE;
--- 174,180 ----
    /* widget_value is a straightforward object translation of emacs's
       Byzantine lisp menu structures */
    wv = xmalloc_widget_value ();
!   wv->name = "Aquamacs/Cocoa";
    wv->value = 0;
    wv->enabled = 1;
    wv->button_type = BUTTON_TYPE_NONE;
***************
*** 1647,1653 ****
    area.size.height= TEXTHEIGHT;
    command = [[[NSTextField alloc] initWithFrame: area] autorelease];
    [[self contentView] addSubview: command];
!   [command setStringValue: @"Emacs"];
    [command setDrawsBackground: NO];
    [command setBezeled: NO];
    [command setSelectable: NO];
--- 1647,1653 ----
    area.size.height= TEXTHEIGHT;
    command = [[[NSTextField alloc] initWithFrame: area] autorelease];
    [[self contentView] addSubview: command];
!   [command setStringValue: @"Aquamacs"];
    [command setDrawsBackground: NO];
    [command setBezeled: NO];
    [command setSelectable: NO];
