*** src/nsterm.h	14 Mar 2009 21:20:08 -0000	1.24
--- src/nsterm.h	15 Mar 2009 20:40:41 -0000
***************
*** 357,363 ****
  #define KEY_NS_NEW_FRAME               ((1<<28)|(0<<16)|12)
  #define KEY_NS_TOGGLE_TOOLBAR          ((1<<28)|(0<<16)|13)
  #define KEY_NS_SHOW_PREFS              ((1<<28)|(0<<16)|14)
! 
  /* could use list to store these, but rest of emacs has a big infrastructure
     for managing a table of bitmap "records" */
  struct ns_bitmap_record
--- 359,366 ----
  #define KEY_NS_NEW_FRAME               ((1<<28)|(0<<16)|12)
  #define KEY_NS_TOGGLE_TOOLBAR          ((1<<28)|(0<<16)|13)
  #define KEY_NS_SHOW_PREFS              ((1<<28)|(0<<16)|14)
! #define KEY_NS_ABOUT                   ((1<<28)|(0<<16)|130)
! #define KEY_NS_CHECK_FOR_UPDATES       ((1<<28)|(0<<16)|131)
  /* could use list to store these, but rest of emacs has a big infrastructure
     for managing a table of bitmap "records" */
  struct ns_bitmap_record
*** src/nsterm.m	14 Mar 2009 21:20:09 -0000	1.67
--- src/nsterm.m	15 Mar 2009 20:40:43 -0000
***************
*** 3896,3939 ****
      /* set up the application menu */
      svcsMenu = [[EmacsMenu alloc] initWithTitle: @"Services"];
      [svcsMenu setAutoenablesItems: NO];
!     appMenu = [[EmacsMenu alloc] initWithTitle: @"Emacs"];
      [appMenu setAutoenablesItems: NO];
      mainMenu = [[EmacsMenu alloc] initWithTitle: @""];
      dockMenu = [[EmacsMenu alloc] initWithTitle: @""];
  
!     [appMenu insertItemWithTitle: @"About Emacs"
!                           action: @selector (orderFrontStandardAboutPanel:)
                     keyEquivalent: @""
                           atIndex: 0];
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 1];
      [appMenu insertItemWithTitle: @"Preferences..."
                            action: @selector (showPreferencesWindow:)
                     keyEquivalent: @","
!                          atIndex: 2];
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 3];
      item = [appMenu insertItemWithTitle: @"Services"
                                   action: @selector (menuDown:)
                            keyEquivalent: @""
!                                 atIndex: 4];
      [appMenu setSubmenu: svcsMenu forItem: item];
  /*    [svcsMenu setSupercell: item]; */
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 5];
!     [appMenu insertItemWithTitle: @"Hide Emacs"
                            action: @selector (hide:)
                     keyEquivalent: @"h"
!                          atIndex: 6];
      item =  [appMenu insertItemWithTitle: @"Hide Others"
                            action: @selector (hideOtherApplications:)
                     keyEquivalent: @"h"
!                          atIndex: 7];
      [item setKeyEquivalentModifierMask: NSCommandKeyMask | NSAlternateKeyMask];
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 8];
!     [appMenu insertItemWithTitle: @"Quit Emacs"
                            action: @selector (terminate:)
                     keyEquivalent: @"q"
!                          atIndex: 9];
  
!     item = [mainMenu insertItemWithTitle: @"Emacs"
                                    action: @selector (menuDown:)
                             keyEquivalent: @""
                                   atIndex: 0];
--- 3940,3987 ----
      /* set up the application menu */
      svcsMenu = [[EmacsMenu alloc] initWithTitle: @"Services"];
      [svcsMenu setAutoenablesItems: NO];
!     appMenu = [[EmacsMenu alloc] initWithTitle: @"Aquamacs"];
      [appMenu setAutoenablesItems: NO];
      mainMenu = [[EmacsMenu alloc] initWithTitle: @""];
      dockMenu = [[EmacsMenu alloc] initWithTitle: @""];
  
!     [appMenu insertItemWithTitle: @"About Aquamacs Emacs"
!                           action: @selector (showAbout:)
                     keyEquivalent: @""
                           atIndex: 0];
!     [appMenu insertItemWithTitle: @"Check for Updates..."
!                           action: @selector (checkForUpdates:)
!                    keyEquivalent: @""
!                          atIndex: 1];
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 2];
      [appMenu insertItemWithTitle: @"Preferences..."
                            action: @selector (showPreferencesWindow:)
                     keyEquivalent: @","
!                          atIndex: 3];
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 4];
      item = [appMenu insertItemWithTitle: @"Services"
                                   action: @selector (menuDown:)
                            keyEquivalent: @""
!                                 atIndex: 5];
      [appMenu setSubmenu: svcsMenu forItem: item];
  /*    [svcsMenu setSupercell: item]; */
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 6];
!     [appMenu insertItemWithTitle: @"Hide Aquamacs Emacs"
                            action: @selector (hide:)
                     keyEquivalent: @"h"
!                          atIndex: 7];
      item =  [appMenu insertItemWithTitle: @"Hide Others"
                            action: @selector (hideOtherApplications:)
                     keyEquivalent: @"h"
!                          atIndex: 8];
      [item setKeyEquivalentModifierMask: NSCommandKeyMask | NSAlternateKeyMask];
!     [appMenu insertItem: [NSMenuItem separatorItem] atIndex: 9];
!     [appMenu insertItemWithTitle: @"Quit Aquamacs Emacs"
                            action: @selector (terminate:)
                     keyEquivalent: @"q"
!                          atIndex: 10];
  
!     item = [mainMenu insertItemWithTitle: @"Aquamacs Emacs"
                                    action: @selector (menuDown:)
                             keyEquivalent: @""
                                   atIndex: 0];
***************
*** 4060,4065 ****
--- 4108,4143 ----
  }
  
  
+ - (void)checkForUpdates: (id)sender
+ {
+   struct frame *emacsframe = SELECTED_FRAME ();
+   NSEvent *theEvent = [NSApp currentEvent];
+ 
+   if (!emacs_event)
+     return;
+   emacs_event->kind = NS_NONKEY_EVENT;
+   emacs_event->code = KEY_NS_CHECK_FOR_UPDATES;
+   emacs_event->modifiers = 0;
+   emacs_event->arg = Qt; /* mark as non-key event */
+ 
+   EV_TRAILER (theEvent);
+ }
+ 
+ - (void)showAbout: (id)sender
+ {
+   struct frame *emacsframe = SELECTED_FRAME ();
+   NSEvent *theEvent = [NSApp currentEvent];
+ 
+   if (!emacs_event)
+     return;
+   emacs_event->kind = NS_NONKEY_EVENT;
+   emacs_event->code = KEY_NS_ABOUT;
+   emacs_event->modifiers = 0;
+   emacs_event->arg = Qt; /* mark as non-key event */
+   EV_TRAILER (theEvent);
+ }
+ 
+ 
  - (void)showPreferencesWindow: (id)sender
  {
    struct frame *emacsframe = SELECTED_FRAME ();
*** src/nsmenu.m	6 Mar 2009 19:06:45 -0000	1.21
--- src/nsmenu.m	15 Mar 2009 20:40:40 -0000
***************
*** 156,162 ****
  
    if (menu == nil)
      {
!       menu = [[EmacsMenu alloc] initWithTitle: @"Emacs"];
        needsSet = YES;
      }
    else
--- 158,164 ----
  
    if (menu == nil)
      {
!       menu = [[EmacsMenu alloc] initWithTitle: @"Aquamacs/Cocoa"];
        needsSet = YES;
      }
    else
***************
*** 174,180 ****
    /* widget_value is a straightforward object translation of emacs's
       Byzantine lisp menu structures */
    wv = xmalloc_widget_value ();
!   wv->name = "Emacs";
    wv->value = 0;
    wv->enabled = 1;
    wv->button_type = BUTTON_TYPE_NONE;
--- 176,182 ----
    /* widget_value is a straightforward object translation of emacs's
       Byzantine lisp menu structures */
    wv = xmalloc_widget_value ();
!   wv->name = "Aquamacs/Cocoa";
    wv->value = 0;
    wv->enabled = 1;
    wv->button_type = BUTTON_TYPE_NONE;
***************
*** 1650,1656 ****
    area.size.height= TEXTHEIGHT;
    command = [[[NSTextField alloc] initWithFrame: area] autorelease];
    [[self contentView] addSubview: command];
!   [command setStringValue: @"Emacs"];
    [command setDrawsBackground: NO];
    [command setBezeled: NO];
    [command setSelectable: NO];
--- 1722,1728 ----
    area.size.height= TEXTHEIGHT;
    command = [[[NSTextField alloc] initWithFrame: area] autorelease];
    [[self contentView] addSubview: command];
!   [command setStringValue: @"Aquamacs"];
    [command setDrawsBackground: NO];
    [command setBezeled: NO];
    [command setSelectable: NO];
*** nextstep/Cocoa/Emacs.base/Contents/Resources/English.lproj/InfoPlist.strings	26 Feb 2009 01:06:53 -0000	1.6
--- nextstep/Cocoa/Emacs.base/Contents/Resources/English.lproj/InfoPlist.strings	15 Mar 2009 20:38:45 -0000
***************
*** 1,6 ****
- /* Localized versions of Info.plist keys */
  
- CFBundleName = "Emacs";
- CFBundleShortVersionString = "Version 23.0.91";
- CFBundleGetInfoString = "Emacs version 23.0.91, NS Windowing";
- NSHumanReadableCopyright = "Copyright (C) 2009 Free Software Foundation, Inc.";
--- 1 ----
