*** ../emacs.temp/lisp/custom.el	2009-01-02 13:59:32.000000000 +1300
--- lisp/custom.el	2009-01-02 13:59:46.000000000 +1300
***************
*** 623,629 ****
  	      ;; We are still loading it when we call this,
  	      ;; and it is not in load-history yet.
  	      ((equal load "cus-edit"))
! 	      (t (condition-case nil (load load) (error nil))))))))
  
  (defvar custom-local-buffer nil
    "Non-nil, in a Customization buffer, means customize a specific buffer.
--- 623,629 ----
  	      ;; We are still loading it when we call this,
  	      ;; and it is not in load-history yet.
  	      ((equal load "cus-edit"))
! 	      (t (condition-case nil (load load nil 'quiet) (error nil))))))))
  
  (defvar custom-local-buffer nil
    "Non-nil, in a Customization buffer, means customize a specific buffer.
