diff -c -r1.315 configure
*** configure	6 Mar 2009 08:09:13 -0000	1.315
--- configure	15 Mar 2009 20:18:56 -0000
***************
*** 2495,2506 ****
      opsys=darwin
      # Define CPP as follows to make autoconf work correctly.
      CPP="${CC-cc} -E -no-cpp-precomp"
-     # Use fink packages if available.
-     if test -d /sw/include && test -d /sw/lib; then
-       GCC_TEST_OPTIONS="-I/sw/include -L/sw/lib"
-       CPP="${CPP} ${GCC_TEST_OPTIONS}"
-       NON_GCC_TEST_OPTIONS=${GCC_TEST_OPTIONS}
-     fi
    ;;
  
    ## HP 9000 series 700 and 800, running HP/UX
--- 2495,2500 ----
