*** ../emacs.temp/configure	2009-01-02 13:59:28.000000000 +1300
--- configure	2009-01-02 13:59:46.000000000 +1300
***************
*** 2503,2514 ****
      opsys=darwin
      # Define CPP as follows to make autoconf work correctly.
      CPP="${CC-cc} -E -no-cpp-precomp"
-     # Use fink packages if available.
-     if test -d /sw/include && test -d /sw/lib; then
-       GCC_TEST_OPTIONS="-I/sw/include -L/sw/lib"
-       CPP="${CPP} ${GCC_TEST_OPTIONS}"
-       NON_GCC_TEST_OPTIONS=${GCC_TEST_OPTIONS}
-     fi
    ;;
  
    ## HP 9000 series 700 and 800, running HP/UX
--- 2503,2508 ----
