*** src/window.c	8 Jan 2009 03:16:08 -0000	1.638
--- src/window.c	15 Mar 2009 20:41:14 -0000
***************
*** 157,162 ****
--- 157,166 ----
  
  static int window_initialized;
  
+ /* List of windows with inhibited header lines  */
+ 
+ Lisp_Object Vheader_line_inhibit_window_list;
+ 
  /* Hook to run when window config changes.  */
  
  static Lisp_Object Qwindow_configuration_change_hook;
***************
*** 5869,5874 ****
--- 5873,5897 ----
    return Fvertical_motion (arg, window);
  }
  
+ /*
+   Return non-nil if the header line in window w is to be inhibited.
+ */
+ 
+ int
+ window_header_line_inhibited_p(w)
+      struct window *w;
+ {
+   Lisp_Object window;
+ 
+   if (!NILP(Vheader_line_inhibit_window_list))
+     {
+       XSETWINDOW (window, w);
+       return (!NILP (Fmember(window, Vheader_line_inhibit_window_list)));
+     }
+   return 0;
+ }
+ 
+ 
  
  
  /***********************************************************************
***************
*** 7164,7169 ****
--- 7187,7195 ----
  
    Qtemp_buffer_show_hook = intern ("temp-buffer-show-hook");
    staticpro (&Qtemp_buffer_show_hook);
+   
+   Vheader_line_inhibit_window_list = Qnil;
+   staticpro (&Vheader_line_inhibit_window_list);
  
    staticpro (&Vwindow_list);
  
***************
*** 7174,7180 ****
    window_scroll_pixel_based_preserve_y = -1;
    window_scroll_preserve_hpos = -1;
    window_scroll_preserve_vpos = -1;
! 
    DEFVAR_LISP ("temp-buffer-show-function", &Vtemp_buffer_show_function,
  	       doc: /* Non-nil means call as function to display a help buffer.
  The function is called with one argument, the buffer to be displayed.
--- 7200,7210 ----
    window_scroll_pixel_based_preserve_y = -1;
    window_scroll_preserve_hpos = -1;
    window_scroll_preserve_vpos = -1;
!  
!   DEFVAR_LISP ("header-line-inhibit-window-list", &Vheader_line_inhibit_window_list,
! 	       doc: /* List of windows in which no header line is shown. */);
!   Vheader_line_inhibit_window_list = Qnil;
!  
    DEFVAR_LISP ("temp-buffer-show-function", &Vtemp_buffer_show_function,
  	       doc: /* Non-nil means call as function to display a help buffer.
  The function is called with one argument, the buffer to be displayed.
*** src/dispextern.h	14 Feb 2009 17:28:27 -0000	1.265
--- src/dispextern.h	15 Mar 2009 20:40:01 -0000
***************
*** 1356,1372 ****
        && !NILP (XBUFFER ((W)->buffer)->mode_line_format)		\
        && WINDOW_TOTAL_LINES (W) > 1)
  
- /* Value is non-zero if window W wants a header line.  */
  
  #define WINDOW_WANTS_HEADER_LINE_P(W)					\
       (!MINI_WINDOW_P ((W))						\
        && !(W)->pseudo_window_p						\
        && FRAME_WANTS_MODELINE_P (XFRAME (WINDOW_FRAME ((W))))		\
        && BUFFERP ((W)->buffer)						\
        && !NILP (XBUFFER ((W)->buffer)->header_line_format)		\
!       && WINDOW_TOTAL_LINES (W) > 1 + !NILP (XBUFFER ((W)->buffer)->mode_line_format))
  
  
  /* Return proper value to be used as baseline offset of font that has
     ASCENT and DESCENT to draw characters by the font at the vertical
     center of the line of frame F.
--- 1356,1377 ----
        && !NILP (XBUFFER ((W)->buffer)->mode_line_format)		\
        && WINDOW_TOTAL_LINES (W) > 1)
  
  
+ extern Lisp_Object Qwindow_wants_header_line_function,
+   Vwindow_wants_header_line_function;
+ 
+ /* Value is non-zero if window W wants a header line.  */
  #define WINDOW_WANTS_HEADER_LINE_P(W)					\
       (!MINI_WINDOW_P ((W))						\
        && !(W)->pseudo_window_p						\
        && FRAME_WANTS_MODELINE_P (XFRAME (WINDOW_FRAME ((W))))		\
        && BUFFERP ((W)->buffer)						\
        && !NILP (XBUFFER ((W)->buffer)->header_line_format)		\
!       && WINDOW_TOTAL_LINES (W) > 1 + !NILP (XBUFFER ((W)->buffer)->mode_line_format) \
!       && !window_header_line_inhibited_p (W))
  
  
+  
  /* Return proper value to be used as baseline offset of font that has
     ASCENT and DESCENT to draw characters by the font at the vertical
     center of the line of frame F.
