*** ../emacs.temp/src/nsterm.m	2009-01-02 13:59:45.000000000 +1300
--- src/nsterm.m	2009-01-02 13:59:45.000000000 +1300
***************
*** 1803,1811 ****
        if (last_mouse_frame && FRAME_LIVE_P (last_mouse_frame))
          f = last_mouse_frame;
        else
!         f = dpyinfo->x_focus_frame ? dpyinfo->x_focus_frame
!                                     : SELECTED_FRAME ();
  
        if (f && f->output_data.ns)  /* TODO: 2nd check no longer needed? */
          {
            view = FRAME_NS_VIEW (*fp);
--- 1806,1820 ----
        if (last_mouse_frame && FRAME_LIVE_P (last_mouse_frame))
          f = last_mouse_frame;
        else
!         {
! 	  /*mouseLoc = [NSEvent mouseLocation];
  
+ CGFloat x = mouseLoc.x; // x location
+ CGFloat y = mouseLoc.y; // y location
+ 	  */
+ 	  f = dpyinfo->x_focus_frame ? dpyinfo->x_focus_frame
+                                     : SELECTED_FRAME ();
+ 	}
        if (f && f->output_data.ns)  /* TODO: 2nd check no longer needed? */
          {
            view = FRAME_NS_VIEW (*fp);
