*** src/nsterm.m	14 Mar 2009 21:20:09 -0000	1.67
--- src/nsterm.m	15 Mar 2009 20:40:43 -0000
***************
*** 1804,1812 ****
        if (last_mouse_frame && FRAME_LIVE_P (last_mouse_frame))
          f = last_mouse_frame;
        else
!         f = dpyinfo->x_focus_frame ? dpyinfo->x_focus_frame
!                                     : SELECTED_FRAME ();
  
        if (f && f->output_data.ns)  /* TODO: 2nd check no longer needed? */
          {
            view = FRAME_NS_VIEW (*fp);
--- 1833,1847 ----
        if (last_mouse_frame && FRAME_LIVE_P (last_mouse_frame))
          f = last_mouse_frame;
        else
!         {
! 	  /*mouseLoc = [NSEvent mouseLocation];
  
+ CGFloat x = mouseLoc.x; // x location
+ CGFloat y = mouseLoc.y; // y location
+ 	  */
+ 	  f = dpyinfo->x_focus_frame ? dpyinfo->x_focus_frame
+                                     : SELECTED_FRAME ();
+ 	}
        if (f && f->output_data.ns)  /* TODO: 2nd check no longer needed? */
          {
            view = FRAME_NS_VIEW (*fp);
