*** ../emacs.temp/lisp/progmodes/python.el	2009-01-02 13:59:35.000000000 +1300
--- lisp/progmodes/python.el	2009-01-02 13:59:47.000000000 +1300
***************
*** 2019,2025 ****
  		 (not (looking-at "[ \t]*$")))
  	    (error "Can't shift all lines enough"))
  	(forward-line))
!       (indent-rigidly start end (- count)))))
  
  (add-to-list 'debug-ignored-errors "^Can't shift all lines enough")
  
--- 2019,2026 ----
  		 (not (looking-at "[ \t]*$")))
  	    (error "Can't shift all lines enough"))
  	(forward-line))
!       (indent-rigidly start end (- count))))
!   (setq deactivate-mark nil))
  
  (add-to-list 'debug-ignored-errors "^Can't shift all lines enough")
  
***************
*** 2034,2040 ****
    (if count
        (setq count (prefix-numeric-value count))
      (setq count python-indent))
!   (indent-rigidly start end count))
  
  (defun python-outline-level ()
    "`outline-level' function for Python mode.
--- 2035,2042 ----
    (if count
        (setq count (prefix-numeric-value count))
      (setq count python-indent))
!   (indent-rigidly start end count)
!   (setq deactivate-mark nil))
  
  (defun python-outline-level ()
    "`outline-level' function for Python mode.
