*** lisp/progmodes/python.el	20 Feb 2009 16:30:56 -0000	1.95
--- lisp/progmodes/python.el	15 Mar 2009 20:33:40 -0000
***************
*** 2023,2029 ****
  		 (not (looking-at "[ \t]*$")))
  	    (error "Can't shift all lines enough"))
  	(forward-line))
!       (indent-rigidly start end (- count)))))
  
  (add-to-list 'debug-ignored-errors "^Can't shift all lines enough")
  
--- 2023,2030 ----
  		 (not (looking-at "[ \t]*$")))
  	    (error "Can't shift all lines enough"))
  	(forward-line))
!       (indent-rigidly start end (- count))))
!   (setq deactivate-mark nil))
  
  (add-to-list 'debug-ignored-errors "^Can't shift all lines enough")
  
***************
*** 2039,2045 ****
    (if count
        (setq count (prefix-numeric-value count))
      (setq count python-indent))
!   (indent-rigidly start end count))
  
  (defun python-outline-level ()
    "`outline-level' function for Python mode.
--- 2040,2047 ----
    (if count
        (setq count (prefix-numeric-value count))
      (setq count python-indent))
!   (indent-rigidly start end count)
!   (setq deactivate-mark nil))
  
  (defun python-outline-level ()
    "`outline-level' function for Python mode.
