*** src/nsterm.m	7 Feb 2009 11:04:07 -0000	1.58
--- src/nsterm.m	2 Mar 2009 00:13:51 -0000
***************
*** 4989,4994 ****
--- 5091,5097 ----
    frameSize.height = FRAME_TEXT_LINES_TO_PIXEL_HEIGHT (emacsframe, rows)
                         + FRAME_NS_TITLEBAR_HEIGHT (emacsframe)
                         + FRAME_NS_TOOLBAR_HEIGHT (emacsframe);
+ #if 0
  #ifdef NS_IMPL_COCOA
    {
      /* this sets window title to have size in it; the wm does this under GS */
***************
*** 5022,5028 ****
        }
    }
  #endif /* NS_IMPL_COCOA */
! /*fprintf (stderr,"    ...size became %.0f x %.0f  (%d x %d)\n",frameSize.width,frameSize.height,cols,rows); */
  
    return frameSize;
  }
--- 5125,5132 ----
        }
    }
  #endif /* NS_IMPL_COCOA */
! #endif /* if 0 */
! fprintf (stderr,"    ...size became %.0f x %.0f  (%d x %d)\n",frameSize.width,frameSize.height,cols,rows); 
  
    return frameSize;
  }
***************
*** 5011,5016 ****
--- 5076,5086 ----
    NSTRACE (windowWillResize);
  /*fprintf (stderr,"Window will resize: %.0f x %.0f\n",frameSize.width,frameSize.height); */
  
+ 
+ /* In Aquamacs, do not quantize, because faces (face-remapping-alist!) will result in 
+    partial lines being displayed at the bottom of windows either way. */
+   return frameSize;
+ 
    cols = FRAME_PIXEL_WIDTH_TO_TEXT_COLS (emacsframe,
  #ifdef NS_IMPL_GNUSTEP
                                          frameSize.width + 3);
***************
*** 5034,5072 ****
    frameSize.height = FRAME_TEXT_LINES_TO_PIXEL_HEIGHT (emacsframe, rows)
                         + FRAME_NS_TITLEBAR_HEIGHT (emacsframe)
                         + FRAME_NS_TOOLBAR_HEIGHT (emacsframe);
- #ifdef NS_IMPL_COCOA
-   {
-     /* this sets window title to have size in it; the wm does this under GS */
-     NSRect r = [[self window] frame];
-     if (r.size.height == frameSize.height && r.size.width == frameSize.width)
-       {
-         if (old_title != 0)
-           {
-             xfree (old_title);
-             old_title = 0;
-           }
-       }
-     else
-       {
-         char *size_title;
-         NSWindow *window = [self window];
-         if (old_title == 0)
-           {
-             const char *t = [[[self window] title] UTF8String];
-             char *pos = strstr (t, "  —  ");
-             if (pos)
-               *pos = '\0';
-             old_title = (char *) xmalloc (strlen (t) + 1);
-             strcpy (old_title, t);
-           }
-         size_title = xmalloc (strlen (old_title) + 40);
-         sprintf (size_title, "%s  —  (%d x %d)", old_title, cols, rows);
-         [window setTitle: [NSString stringWithUTF8String: size_title]];
-         [window display];
-         xfree (size_title);
-       }
-   }
- #endif /* NS_IMPL_COCOA */
  /*fprintf (stderr,"    ...size became %.0f x %.0f  (%d x %d)\n",frameSize.width,frameSize.height,cols,rows); */
  
    return frameSize;
--- 5099,5104 ----
***************
*** 5043,5060 ****
    NSTRACE (windowDidResize);
  /*fprintf (stderr,"windowDidResize: %.0f\n",[theWindow frame].size.height); */
  
- #ifdef NS_IMPL_COCOA
-   if (old_title != 0)
-     {
-       xfree (old_title);
-       old_title = 0;
-     }
- #endif /* NS_IMPL_COCOA */
- 
    if (cols > 0 && rows > 0)
      x_set_window_size (emacsframe, 0, cols, rows);
  
    ns_send_appdefined (-1);
  }
  
  
--- 5146,5159 ----
    NSTRACE (windowDidResize);
  /*fprintf (stderr,"windowDidResize: %.0f\n",[theWindow frame].size.height); */
  
    if (cols > 0 && rows > 0)
      x_set_window_size (emacsframe, 0, cols, rows);
  
    ns_send_appdefined (-1);
+ 
+ #ifdef NS_IMPL_COCOA
+   NSEnableScreenUpdates ();
+ #endif
  }
  
  
***************
*** 5088,5101 ****
    NSTRACE (windowDidResize);
  /*fprintf (stderr,"windowDidResize: %.0f\n",[theWindow frame].size.height); */
  
- #ifdef NS_IMPL_COCOA
-   if (old_title != 0)
-     {
-       xfree (old_title);
-       old_title = 0;
-     }
- #endif /* NS_IMPL_COCOA */
- 
    if (cols > 0 && rows > 0)
      x_set_window_size (emacsframe, 0, cols, rows);
  
--- 5125,5130 ----
