*** ../emacs.temp/lisp/startup.el	2009-03-19 16:13:34.000000000 +1300
--- lisp/startup.el	2009-03-19 16:13:35.000000000 +1300
***************
*** 1041,1046 ****
--- 1041,1048 ----
  		      ;; into user-init-file.
  		      (setq user-init-file t)
  		      (load user-init-file-1 t t)
+ 		      (if (fboundp 'aquamacs-load-preferences) 
+ 			  (aquamacs-load-preferences))
  
  		      (when (eq user-init-file t)
  			;; If we did not find ~/.emacs, try
