*** ../emacs.temp/lisp/isearch.el	2009-01-02 13:59:34.000000000 +1300
--- lisp/isearch.el	2009-01-02 13:59:46.000000000 +1300
***************
*** 1564,1569 ****
--- 1564,1572 ----
  	    ;; If the string was found but was completely invisible,
  	    ;; it might now be partly visible, so try again.
  	    (prog1 isearch-hidden (setq isearch-hidden nil)))
+     (if (and transient-mark-mode 
+ 	     (not (and cua-mode cua--explicit-region-start)))
+ 	(deactivate-mark))
      ;; In reverse search, adding stuff at
      ;; the end may cause zero or many more chars to be
      ;; matched, in the string following point.
***************
*** 1594,1601 ****
  		       (min isearch-opoint
  			    isearch-barrier
  			    (1+ isearch-other-end)))))
!       (isearch-search)
!       ))
    (isearch-push-state)
    (if isearch-op-fun (funcall isearch-op-fun))
    (isearch-update))
--- 1597,1603 ----
  		       (min isearch-opoint
  			    isearch-barrier
  			    (1+ isearch-other-end)))))
!       (isearch-search)))
    (isearch-push-state)
    (if isearch-op-fun (funcall isearch-op-fun))
    (isearch-update))
*** ../emacs.temp/lisp/menu-bar.el	2009-01-02 13:59:46.000000000 +1300
--- lisp/menu-bar.el	2009-01-02 13:59:46.000000000 +1300
***************
*** 540,553 ****
  (defun clipboard-kill-ring-save (beg end)
    "Copy region to kill ring, and save in the X clipboard."
    (interactive "r")
!   (let ((x-select-enable-clipboard t))
!     (kill-ring-save beg end)))
  
  (defun clipboard-kill-region (beg end)
    "Kill the region, and save it in the X clipboard."
    (interactive "r")
!   (let ((x-select-enable-clipboard t))
!     (kill-region beg end)))
  
  (defun menu-bar-enable-clipboard ()
    "Make CUT, PASTE and COPY (keys and menu bar items) use the clipboard.
--- 540,555 ----
  (defun clipboard-kill-ring-save (beg end)
    "Copy region to kill ring, and save in the X clipboard."
    (interactive "r")
!   (when (or (not transient-mark-mode) mark-active)
!     (let ((x-select-enable-clipboard t))
!       (kill-ring-save beg end))))
  
  (defun clipboard-kill-region (beg end)
    "Kill the region, and save it in the X clipboard."
    (interactive "r")
!   (when (or (not transient-mark-mode) mark-active)
!     (let ((x-select-enable-clipboard t))
!       (kill-region beg end))))
  
  (defun menu-bar-enable-clipboard ()
    "Make CUT, PASTE and COPY (keys and menu bar items) use the clipboard.
*** ../emacs.temp/lisp/mouse.el	2009-01-02 13:59:34.000000000 +1300
--- lisp/mouse.el	2009-01-02 13:59:46.000000000 +1300
***************
*** 955,961 ****
      (mouse-move-drag-overlay mouse-drag-overlay start-point start-point
                               click-count)
      (overlay-put mouse-drag-overlay 'window start-window)
!     (deactivate-mark)
      (let (event end end-point last-end-point)
        (track-mouse
  	(while (progn
--- 955,962 ----
      (mouse-move-drag-overlay mouse-drag-overlay start-point start-point
                               click-count)
      (overlay-put mouse-drag-overlay 'window start-window)
!     (unless (and cua-mode cua--explicit-region-start)
!       (deactivate-mark))
      (let (event end end-point last-end-point)
        (track-mouse
  	(while (progn
