*** lisp/menu-bar.el	Sat Apr 21 19:21:34 2007
--- lisp/menu-bar.el	Sat Apr 21 19:20:42 2007
***************
*** 1161,1167 ****
    '("--"))

  (define-key menu-bar-tools-menu [pcl-cvs]
!   '(menu-item "PCL-CVS" cvs-global-menu))
  (define-key menu-bar-tools-menu [vc] nil) ;Create the place for the VC menu.

  (define-key menu-bar-tools-menu [separator-compare]
--- 1191,1197 ----
    '("--"))

  (define-key menu-bar-tools-menu [pcl-cvs]
!   '(menu-item "CVS Repositories" cvs-global-menu))
  (define-key menu-bar-tools-menu [vc] nil) ;Create the place for the VC menu.

  (define-key menu-bar-tools-menu [separator-compare]

  
