*** ../emacs.temp/lisp/menu-bar.el	2009-01-02 13:59:46.000000000 +1300
--- lisp/menu-bar.el	2009-01-02 13:59:46.000000000 +1300
***************
*** 1324,1330 ****
    '("--"))
  
  (define-key menu-bar-tools-menu [pcl-cvs]
!   '(menu-item "PCL-CVS" cvs-global-menu))
  (define-key menu-bar-tools-menu [vc] nil) ;Create the place for the VC menu.
  
  (define-key menu-bar-tools-menu [separator-compare]
--- 1324,1330 ----
    '("--"))
  
  (define-key menu-bar-tools-menu [pcl-cvs]
!   '(menu-item "CVS Repositories" cvs-global-menu))
  (define-key menu-bar-tools-menu [vc] nil) ;Create the place for the VC menu.
  
  (define-key menu-bar-tools-menu [separator-compare]
