*** ../emacs.temp/lisp/window.el	2009-03-19 16:13:33.000000000 +1300
--- lisp/window.el	2009-03-19 16:13:36.000000000 +1300
***************
*** 1487,1515 ****
  	(fit-window-to-buffer window (window-height window)))))
  
  (defun kill-buffer-and-window ()
!   "Kill the current buffer and delete the selected window."
!   (interactive)
!   (let ((window-to-delete (selected-window))
  	(buffer-to-kill (current-buffer))
  	(delete-window-hook (lambda ()
  			      (condition-case nil
! 				  (delete-window)
  				(error nil)))))
!     (unwind-protect
  	(progn
  	  (add-hook 'kill-buffer-hook delete-window-hook t t)
  	  (if (kill-buffer (current-buffer))
  	      ;; If `delete-window' failed before, we rerun it to regenerate
  	      ;; the error so it can be seen in the echo area.
  	      (when (eq (selected-window) window-to-delete)
! 		(delete-window))))
!       ;; If the buffer is not dead for some reason (probably because
!       ;; of a `quit' signal), remove the hook again.
!       (condition-case nil
  	  (with-current-buffer buffer-to-kill
  	    (remove-hook 'kill-buffer-hook delete-window-hook t))
  	(error nil)))))
  
  (defun quit-window (&optional kill window)
    "Quit WINDOW and bury its buffer.
  With a prefix argument, kill the buffer instead.  WINDOW defaults
--- 1487,1518 ----
  	(fit-window-to-buffer window (window-height window)))))
  
  (defun kill-buffer-and-window ()
!  "Kill the current buffer and delete the selected window."
!  (interactive)
!  (let ((window-to-delete (selected-window))
  	(buffer-to-kill (current-buffer))
  	(delete-window-hook (lambda ()
  			      (condition-case nil
! 				  (if (eq (window-buffer) buffer-to-kill)
! 				      (delete-window))
  				(error nil)))))
!    (unwind-protect
  	(progn
  	  (add-hook 'kill-buffer-hook delete-window-hook t t)
  	  (if (kill-buffer (current-buffer))
  	      ;; If `delete-window' failed before, we rerun it to regenerate
  	      ;; the error so it can be seen in the echo area.
  	      (when (eq (selected-window) window-to-delete)
! 		(if (eq (window-buffer) buffer-to-kill)
! 		    (delete-window)))))
!      ;; If the buffer is not dead for some reason (probably because
!      ;; of a `quit' signal), remove the hook again.
!      (condition-case nil
  	  (with-current-buffer buffer-to-kill
  	    (remove-hook 'kill-buffer-hook delete-window-hook t))
  	(error nil)))))
  
+ 
  (defun quit-window (&optional kill window)
    "Quit WINDOW and bury its buffer.
  With a prefix argument, kill the buffer instead.  WINDOW defaults
