*** src/mactoolbox.c	6 Apr 2008 01:54:44 -0000	1.1.2.1
--- src/mactoolbox.c	6 Apr 2008 13:00:25 -0000
***************
*** 4773,4780 ****
        SetMenuItemCommandKey (menu, menu_index, false, 0);
        InsertMenuItemTextWithCFString (menu, NULL,
  				      0, kMenuItemAttrSeparator, 0);
!       InsertMenuItemTextWithCFString (menu, CFSTR ("About Emacs"),
  				      0, 0, kHICommandAbout);
      }
  #else	/* !MAC_OSX */
  #if TARGET_API_MAC_CARBON
--- 4778,4788 ----
        SetMenuItemCommandKey (menu, menu_index, false, 0);
        InsertMenuItemTextWithCFString (menu, NULL,
  				      0, kMenuItemAttrSeparator, 0);
!       InsertMenuItemTextWithCFString (menu, CFSTR ("About Aquamacs Emacs"),
  				      0, 0, kHICommandAbout);
+       InsertMenuItemTextWithCFString (menu, CFSTR ("Check for Updates..."),
+ 				      1, 0, 'chku');
+ 
      }
  #else	/* !MAC_OSX */
  #if TARGET_API_MAC_CARBON
*** lisp/startup.el	6 Mar 2008 00:08:01 -0000	1.436.2.16
--- lisp/startup.el	1 Jun 2008 12:40:50 -0000
***************
*** 1513,1519 ****
        (message "%s" (startup-echo-area-message))
        (setq buffer-read-only t)
        (goto-char (point-min))
!       (forward-line 3))))
  
  (defun fancy-splash-frame ()
    "Return the frame to use for the fancy splash screen.
--- 1516,1522 ----
        (message "%s" (startup-echo-area-message))
        (setq buffer-read-only t)
        (goto-char (point-min))
!       (forward-line 1))))
  
  (defun fancy-splash-frame ()
    "Return the frame to use for the fancy splash screen.
