*** src/macterm.c	Wed Sep 28 00:40:52 2005
--- src/macterm.c	Wed Sep 28 00:41:16 2005
***************
*** 12691,12697 ****
        SetMenuItemCommandKey (menu, menu_index, false, 0);
        InsertMenuItemTextWithCFString (menu, NULL,
  				      0, kMenuItemAttrSeparator, 0);
!       InsertMenuItemTextWithCFString (menu, CFSTR ("About Emacs"),
  				      0, 0, kHICommandAbout);
      }
  #else	/* !MAC_OSX */
--- 12734,12740 ----
        SetMenuItemCommandKey (menu, menu_index, false, 0);
        InsertMenuItemTextWithCFString (menu, NULL,
  				      0, kMenuItemAttrSeparator, 0);
!       InsertMenuItemTextWithCFString (menu, CFSTR ("About Aquamacs Emacs"),
  				      0, 0, kHICommandAbout);
      }
  #else	/* !MAC_OSX */
