*** toolbar-x.el	25 Feb 2008 17:42:19 +0000	1.7
--- toolbar-x.el	28 Feb 2008 18:40:21 +0000	
***************
*** 124,130 ****
  			      ;;(file-directory-p x)
  			      x))
  		     load-path))
!    (list data-directory))
    "List of directories where toolbarx finds its images.")
  
  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
--- 124,130 ----
  			      ;;(file-directory-p x)
  			      x))
  		     load-path))
!    (list (concat data-directory "images/")))
    "List of directories where toolbarx finds its images.")
  
  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
***************
*** 502,511 ****
  					    ; (defined with `defimage')
  			      (consp (eval val))
  			      (eq (car (eval val)) 'image))
! 			 (and (listp val) ; or list with 4 strings or
! 					  ; image descriptors
  			      (= (length val) 4)
! 			      (dolist (i val all-obj-ok)
  				(setq all-obj-ok
  				      (and all-obj-ok
  					   (or (stringp i)
--- 502,511 ----
  					    ; (defined with `defimage')
  			      (consp (eval val))
  			      (eq (car (eval val)) 'image))
! 			 (and (sequencep val) ; or list or vector with 4 strings or
! 					; image descriptors
  			      (= (length val) 4)
! 			      (dolist (i (append val nil) all-obj-ok)
  				(setq all-obj-ok
  				      (and all-obj-ok
  					   (or (stringp i)
***************
*** 1121,1134 ****
    ;; following should hopefully get us to all images ultimately.
  
    (let ((file))
!     (dolist (i '("" ".xpm" ".xbm" ".pbm"))
        (unless file
  	(setq file (locate-library (concat image i) t toolbarx-image-path))))
      (if (featurep 'xemacs)
  	(and file (make-glyph file))
        (if file
  	  (create-image file)
! 	(find-image `((:type xpm :file ,(concat image ".xpm"))
  		      (:type xbm :file ,(concat image ".xbm"))
  		      (:type pbm :file ,(concat image ".pbm"))))))))
  
--- 1121,1135 ----
    ;; following should hopefully get us to all images ultimately.
  
    (let ((file))
!     (dolist (i '("" ".png" ".xpm" ".xbm" ".pbm"))
        (unless file
  	(setq file (locate-library (concat image i) t toolbarx-image-path))))
      (if (featurep 'xemacs)
  	(and file (make-glyph file))
        (if file
  	  (create-image file)
! 	(find-image `((:type png :file ,(concat image ".png"))
! 		      (:type xpm :file ,(concat image ".xpm"))
  		      (:type xbm :file ,(concat image ".xbm"))
  		      (:type pbm :file ,(concat image ".pbm"))))))))
  
*** toolbar-x.el	27 Apr 2008 08:31:43 +0100	1.11
--- toolbar-x.el	27 Apr 2008 21:01:02 +0100	
***************
*** 630,635 ****
--- 630,636 ----
    (let* ((props-types-alist
  	  '((:image	      toolbarx-test-image-type)
  	    (:command	      toolbarx-test-any-type)
+ 	    (:title	      toolbarx-test-string-or-nil)
  	    (:enable	      toolbarx-test-any-type)
  	    (:visible	      toolbarx-test-any-type)
  	    (:help	      toolbarx-test-string-or-nil)
***************
*** 1240,1246 ****
  			     (cadr (memq :button filtered-props))))
  	       (menuitem (append
  			  (list 'menu-item
! 				(toolbarx-make-string-from-symbol symbol)
  				command
  				:image image-descriptor)
  			  (when (car help)
--- 1241,1248 ----
  			     (cadr (memq :button filtered-props))))
  	       (menuitem (append
  			  (list 'menu-item
! 				(or (cadr (memq :title filtered-props)) 
! 				    (toolbarx-make-string-from-symbol symbol))
  				command
  				:image image-descriptor)
  			  (when (car help)
