Index: src/macfns.c
===================================================================
RCS file: /cvsroot/emacs/emacs/src/macfns.c,v
retrieving revision 1.63
diff -c -r1.63 macfns.c
*** src/macfns.c	4 Jul 2005 16:06:33 -0000	1.63
--- src/macfns.c	17 Jul 2005 20:28:16 -0000
***************
*** 3028,3033 ****
--- 3029,3059 ----
    return make_number ((int) (dpyinfo->width * 25.4 / dpyinfo->resx));
  }
  
+ 
+ 
+ DEFUN ("mac-display-available-pixel-bounds", Fmac_display_available_pixel_bounds, Smac_display_available_pixel_bounds, 0, 0, 0,
+        doc: /* Get the available bounds of the main display in pixels, 
+ excluding the Dock. Return a 4 element list with the bounds
+ in the form of (left top right bottom). */)
+   (void)
+ {
+   Rect rect;
+   BLOCK_INPUT;   
+   GetAvailableWindowPositioningBounds(LMGetMainDevice(), &rect);
+ 
+   Lisp_Object rv = 
+     Fcons( make_number (rect.left),
+ 	   Fcons( make_number (rect.top),
+ 		  Fcons( make_number (rect.right),
+ 			 Fcons( make_number (rect.bottom), Qnil))));
+   UNBLOCK_INPUT;
+   return rv;
+ 
+ }
+ 
+ 
+ 
+ 
  DEFUN ("x-display-backing-store", Fx_display_backing_store,
         Sx_display_backing_store, 0, 1, 0,
         doc: /* Returns an indication of whether display DISPLAY does backing store.
***************
*** 4586,4591 ****
--- 4611,4618 ----
    defsubr (&Sx_close_connection);
    defsubr (&Sx_display_list);
    defsubr (&Sx_synchronize);
+ 
+   defsubr (&Smac_display_available_pixel_bounds);
  
    /* Setting callback functions for fontset handler.  */
    get_font_info_func = x_get_font_info;
