Index: lisp/startup.el
===================================================================
RCS file: /cvsroot/emacs/emacs/lisp/startup.el,v
retrieving revision 1.368
diff -c -r1.368 startup.el
*** lisp/startup.el	15 Jul 2005 14:47:06 -0000	1.368
--- lisp/startup.el	17 Jul 2005 22:07:56 -0000
***************
*** 463,469 ****
  		  (if (string-match "^\\(xterm\\|rxvt\\|dtterm\\|eterm\\)"
  				    term)
  		      (setq default-frame-background-mode 'light)))
! 		(frame-set-background-mode (selected-frame)))))
  
  	;; Now we know the user's default font, so add it to the menu.
  	(if (fboundp 'font-menu-add-default)
--- 463,472 ----
  		  (if (string-match "^\\(xterm\\|rxvt\\|dtterm\\|eterm\\)"
  				    term)
  		      (setq default-frame-background-mode 'light)))
! 		(frame-set-background-mode (selected-frame))))
! 
! 	  ;; time to make the frame visible (Aquamacs)
! 	  (make-frame-visible))
  
  	;; Now we know the user's default font, so add it to the menu.
  	(if (fboundp 'font-menu-add-default)
***************
*** 721,726 ****
--- 724,732 ----
      (and command-line-args
           (setcdr command-line-args args)))
  
+   ;; the initial frame is hidden in Aquamacs
+   (setq initial-frame-alist (cons '(visibility . nil) initial-frame-alist))
+ 
    ;; Under X Window, this creates the X frame and deletes the terminal frame.
    (when (fboundp 'frame-initialize)
      (frame-initialize))
***************
*** 2146,2151 ****
--- 2119,2131 ----
  	   (run-hooks 'term-setup-hook))
        (setq inhibit-startup-hooks t)
  
+       ;; time to make the frame visible (Aquamacs)
+       ;; done before calling `frame-notice-user-settings'
+       ;; as a workaround because the modificiation of 
+       ;; frame position does not appear to work while the
+       ;; the frame is hidden.  dr. 04/2008
+       (make-frame-visible) 
+ 
        ;; It's important to notice the user settings before we
        ;; display the startup message; otherwise, the settings
        ;; won't take effect until the user gives the first
