Index: lisp/cus-edit.el
===================================================================
RCS file: /sources/emacs/emacs/lisp/cus-edit.el,v
retrieving revision 1.319
diff -c -r1.319 cus-edit.el
*** lisp/cus-edit.el	13 Apr 2007 09:05:43 -0000	1.319
--- lisp/cus-edit.el	13 Jul 2007 20:47:28 -0000
***************
*** 1041,1047 ****
    (let ((name (format "*Customize Group: %s*"
  		      (custom-unlispify-tag-name group))))
      (if (get-buffer name)
! 	(pop-to-buffer name)
        (custom-buffer-create (list (list group 'custom-group))
  			    name
  			    (concat " for group "
--- 1041,1047 ----
    (let ((name (format "*Customize Group: %s*"
  		      (custom-unlispify-tag-name group))))
      (if (get-buffer name)
! 	(switch-to-buffer name)
        (custom-buffer-create (list (list group 'custom-group))
  			    name
  			    (concat " for group "
***************
*** 1483,1489 ****
  OPTIONS should be an alist of the form ((SYMBOL WIDGET)...), where
  SYMBOL is a customization option, and WIDGET is a widget for editing
  that option."
!   (pop-to-buffer (custom-get-fresh-buffer (or name "*Customization*")))
    (custom-buffer-create-internal options description))
  
  ;;;###autoload
--- 1483,1489 ----
  OPTIONS should be an alist of the form ((SYMBOL WIDGET)...), where
  SYMBOL is a customization option, and WIDGET is a widget for editing
  that option."
!   (switch-to-buffer (custom-get-fresh-buffer (or name "*Customization*")))
    (custom-buffer-create-internal options description))
  
  ;;;###autoload
