*** lisp/custom.el	20 Apr 2006 14:28:31 +0100	1.126
--- lisp/custom.el	21 Apr 2006 15:12:09 +0100	
***************
*** 608,614 ****
  	      ;; We are still loading it when we call this,
  	      ;; and it is not in load-history yet.
  	      ((equal load "cus-edit"))
! 	      (t (condition-case nil (load load) (error nil))))))))
  
  (defvar custom-local-buffer nil
    "Non-nil, in a Customization buffer, means customize a specific buffer.
--- 608,614 ----
  	      ;; We are still loading it when we call this,
  	      ;; and it is not in load-history yet.
  	      ((equal load "cus-edit"))
! 	      (t (condition-case nil (load load nil 'quiet) (error nil))))))))
  
  (defvar custom-local-buffer nil
    "Non-nil, in a Customization buffer, means customize a specific buffer.
