*** lisp/files.el	5 Aug 2008 21:44:43 -0000	1.896.2.42
--- lisp/files.el	10 Feb 2009 01:50:25 -0000
***************
*** 4582,4589 ****
  		     ;; Try a few alternative directories, to get one we can
  		     ;; write it.
  		     (cond
- 		      ((file-writable-p default-directory) default-directory)
  		      ((file-writable-p "/var/tmp/") "/var/tmp/")
  		      ("~/")))))
  	       (if (and (memq system-type '(ms-dos windows-nt cygwin))
  			;; Don't modify remote (ange-ftp) filenames
--- 4619,4626 ----
  		     ;; Try a few alternative directories, to get one we can
  		     ;; write it.
  		     (cond
  		      ((file-writable-p "/var/tmp/") "/var/tmp/")
+ 		      ((file-writable-p "/tmp/") "/tmp/")
  		      ("~/")))))
  	       (if (and (memq system-type '(ms-dos windows-nt cygwin))
  			;; Don't modify remote (ange-ftp) filenames
