*** src/window.c	12 Sep 2006 10:58:06 +0100	1.557
--- src/window.c	13 Dec 2006 09:39:10 +0000	
***************
*** 54,59 ****
--- 54,61 ----
  Lisp_Object Qwindow_size_fixed;
  extern Lisp_Object Qleft_margin, Qright_margin;
  
+ Lisp_Object Qselect;
+ 
  static int displayed_window_lines P_ ((struct window *));
  static struct window *decode_window P_ ((Lisp_Object));
  static int count_windows P_ ((struct window *));
***************
*** 3390,3396 ****
        if (FRAME_ICONIFIED_P (f))
  	Fmake_frame_visible (frame);
        else if (FRAME_VISIBLE_P (f))
! 	Fraise_frame (frame);
      }
  
    return window;
--- 3392,3403 ----
        if (FRAME_ICONIFIED_P (f))
  	Fmake_frame_visible (frame);
        else if (FRAME_VISIBLE_P (f))
! 	{
! 	  if (EQ (display_buffer_reuse_frames, Qselect))
! 	    Fselect_frame (frame);
! 	  else
! 	    Fraise_frame (frame);
! 	}
      }
  
    return window;
***************
*** 7133,7138 ****
--- 7140,7148 ----
  void
  syms_of_window ()
  {
+   Qselect = intern ("select");
+   staticpro (&Qselect);
+ 
    Qscroll_up = intern ("scroll-up");
    staticpro (&Qscroll_up);
  
***************
*** 7213,7219 ****
  
    DEFVAR_BOOL ("display-buffer-reuse-frames", &display_buffer_reuse_frames,
  	       doc: /* *Non-nil means `display-buffer' should reuse frames.
! If the buffer in question is already displayed in a frame, raise that frame.  */);
    display_buffer_reuse_frames = 0;
  
    DEFVAR_LISP ("pop-up-frame-function", &Vpop_up_frame_function,
--- 7223,7230 ----
  
    DEFVAR_BOOL ("display-buffer-reuse-frames", &display_buffer_reuse_frames,
  	       doc: /* *Non-nil means `display-buffer' should reuse frames.
! If the buffer in question is already displayed in a frame, raise that frame.  
! `select' means that the frame is not only raised, but also selected.*/);
    display_buffer_reuse_frames = 0;
  
    DEFVAR_LISP ("pop-up-frame-function", &Vpop_up_frame_function,
