*** lisp/ediff-wind.el	25 Jul 2007 05:47:27 +0100	1.47.2.1
--- lisp/ediff-wind.el	29 Sep 2007 12:07:57 +0100	
***************
*** 952,960 ****
  	  adjusted-parameters
  	  (list
  	   ;; possibly change surrogate minibuffer
! 	   (cons 'minibuffer
! 		 (minibuffer-window
! 		  designated-minibuffer-frame))
  	   (cons 'width fwidth)
  	   (cons 'height fheight)
  	   (cons 'user-position t)
--- 952,961 ----
  	  adjusted-parameters
  	  (list
  	   ;; possibly change surrogate minibuffer
! ;; Workaround: disabled in Aquamacs because it leads to an error
! ;; 	   (cons 'minibuffer
! ;; 		 (minibuffer-window
! ;; 		  designated-minibuffer-frame))
  	   (cons 'width fwidth)
  	   (cons 'height fheight)
  	   (cons 'user-position t)
