*** lisp/faces.el	13 Apr 2007 18:24:53 +0100	1.370
--- lisp/faces.el	13 Jul 2007 18:56:06 +0100	
***************
*** 320,326 ****
  
  (defun set-face-attributes-from-resources (face frame)
    "Set attributes of FACE from X resources for FRAME."
!   (when (memq (framep frame) '(x w32 mac))
      (dolist (definition face-x-resources)
        (let ((attribute (car definition)))
  	(dolist (entry (cdr definition))
--- 320,326 ----
  
  (defun set-face-attributes-from-resources (face frame)
    "Set attributes of FACE from X resources for FRAME."
!   (when (eq (framep frame) 'x)
      (dolist (definition face-x-resources)
        (let ((attribute (car definition)))
  	(dolist (entry (cdr definition))
***************
*** 1789,1796 ****
        (condition-case ()
  	  (progn
  	    (face-spec-set face (face-user-default-spec face) frame)
! 	    (if (memq window-system '(x w32 mac))
! 		(make-face-x-resource-internal face frame))
  	    (internal-merge-in-global-face face frame))
  	(error nil)))
      ;; Apply the attributes specified by frame parameters.  This
--- 1789,1796 ----
        (condition-case ()
  	  (progn
  	    (face-spec-set face (face-user-default-spec face) frame)
! 	     (if (eq window-system 'x)
!  		(make-face-x-resource-internal face frame))
  	    (internal-merge-in-global-face face frame))
  	(error nil)))
      ;; Apply the attributes specified by frame parameters.  This
