*** lisp/textmodes/flyspell.el	Wed Dec 31 18:00:00 1969
--- lisp/textmodes/flyspell.el	Tue May 13 12:05:09 2008
***************
*** 1006,1011 ****
--- 1006,1012 ----
    (interactive (list ispell-following-word))
    (save-excursion
      ;; use the correct dictionary
+     (ispell-maybe-find-aspell-dictionaries)
      (flyspell-accept-buffer-local-defs)
      (let* ((cursor-location (point))
             (flyspell-word (flyspell-get-word following))
***************
*** 1506,1511 ****
--- 1507,1513 ----
      (setq flyspell-large-region-buffer curbuf)
      (setq flyspell-large-region-beg beg)
      (setq flyspell-large-region-end end)
+     (ispell-maybe-find-aspell-dictionaries)
      (flyspell-accept-buffer-local-defs)
      (set-buffer buffer)
      (erase-buffer)
