Index: src/window.c
===================================================================
RCS file: /sources/emacs/emacs/src/window.c,v
retrieving revision 1.575.2.7
diff -c -r1.575.2.7 window.c
*** src/window.c	8 Jan 2008 04:29:43 -0000	1.575.2.7
--- src/window.c	1 Mar 2008 15:00:37 -0000
***************
*** 203,208 ****
--- 203,212 ----
  
  static int window_initialized;
  
+ /* List of windows with inhibited header lines  */
+ 
+ Lisp_Object Vheader_line_inhibit_window_list;
+ 
  /* Hook to run when window config changes.  */
  
  Lisp_Object Qwindow_configuration_change_hook;
***************
*** 6012,6017 ****
--- 6016,6040 ----
    return Fvertical_motion (arg, window);
  }
  
+ /*
+   Return non-nil if the header line in window w is to be inhibited.
+ */
+ 
+ int
+ window_header_line_inhibited_p(w)
+      struct window *w;
+ {
+   Lisp_Object window;
+ 
+   if (!NILP(Vheader_line_inhibit_window_list))
+     {
+       XSETWINDOW (window, w);
+       return (!NILP (Fmember(window, Vheader_line_inhibit_window_list)));
+     }
+   return 0;
+ }
+ 
+ 
  
  
  /***********************************************************************
***************
*** 7316,7321 ****
--- 7339,7347 ----
  
    Qtemp_buffer_show_hook = intern ("temp-buffer-show-hook");
    staticpro (&Qtemp_buffer_show_hook);
+   
+   Vheader_line_inhibit_window_list = Qnil;
+   staticpro (&Vheader_line_inhibit_window_list);
  
    staticpro (&Vwindow_list);
  
***************
*** 7332,7337 ****
--- 7358,7369 ----
  the buffer; `temp-buffer-show-hook' is not run unless this function runs it.  */);
    Vtemp_buffer_show_function = Qnil;
  
+ 
+   DEFVAR_LISP ("header-line-inhibit-window-list", &Vheader_line_inhibit_window_list,
+     doc: /* List of windows in which no header line is shown. */);
+   Vheader_line_inhibit_window_list = Qnil;
+ 
+ 
    DEFVAR_LISP ("display-buffer-function", &Vdisplay_buffer_function,
  	       doc: /* If non-nil, function to call to handle `display-buffer'.
  It will receive two args, the buffer and a flag which if non-nil means
Index: src/dispextern.h
===================================================================
RCS file: /sources/emacs/emacs/src/dispextern.h,v
retrieving revision 1.225.2.7
diff -c -r1.225.2.7 dispextern.h
*** src/dispextern.h	8 Jan 2008 04:30:14 -0000	1.225.2.7
--- src/dispextern.h	1 Mar 2008 15:00:37 -0000
***************
*** 1321,1337 ****
        && !NILP (XBUFFER ((W)->buffer)->mode_line_format)		\
        && WINDOW_TOTAL_LINES (W) > 1)
  
- /* Value is non-zero if window W wants a header line.  */
  
  #define WINDOW_WANTS_HEADER_LINE_P(W)					\
       (!MINI_WINDOW_P ((W))						\
        && !(W)->pseudo_window_p						\
        && FRAME_WANTS_MODELINE_P (XFRAME (WINDOW_FRAME ((W))))		\
        && BUFFERP ((W)->buffer)						\
        && !NILP (XBUFFER ((W)->buffer)->header_line_format)		\
!       && WINDOW_TOTAL_LINES (W) > 1 + !NILP (XBUFFER ((W)->buffer)->mode_line_format))
  
  
  /* Return proper value to be used as baseline offset of font that has
     ASCENT and DESCENT to draw characters by the font at the vertical
     center of the line of frame F.
--- 1321,1342 ----
        && !NILP (XBUFFER ((W)->buffer)->mode_line_format)		\
        && WINDOW_TOTAL_LINES (W) > 1)
  
  
+ extern Lisp_Object Qwindow_wants_header_line_function,
+   Vwindow_wants_header_line_function;
+ 
+ /* Value is non-zero if window W wants a header line.  */
  #define WINDOW_WANTS_HEADER_LINE_P(W)					\
       (!MINI_WINDOW_P ((W))						\
        && !(W)->pseudo_window_p						\
        && FRAME_WANTS_MODELINE_P (XFRAME (WINDOW_FRAME ((W))))		\
        && BUFFERP ((W)->buffer)						\
        && !NILP (XBUFFER ((W)->buffer)->header_line_format)		\
!       && WINDOW_TOTAL_LINES (W) > 1 + !NILP (XBUFFER ((W)->buffer)->mode_line_format) \
!       && !window_header_line_inhibited_p (W))
  
  
+  
  /* Return proper value to be used as baseline offset of font that has
     ASCENT and DESCENT to draw characters by the font at the vertical
     center of the line of frame F.
*** lisp/info.el	06 Apr 2008 07:51:37 -0400	1.501.2.7
--- lisp/info.el	17 Oct 2008 22:46:43 -0400	
***************
*** 3454,3463 ****
    (make-local-variable 'Info-history)
    (make-local-variable 'Info-history-forward)
    (make-local-variable 'Info-index-alternatives)
!   (setq header-line-format
! 	(if Info-use-header-line
! 	    '(:eval (get-text-property (point-min) 'header-line))
! 	  nil)) ; so the header line isn't displayed
    (set (make-local-variable 'tool-bar-map) info-tool-bar-map)
    ;; This is for the sake of the invisible text we use handling titles.
    (make-local-variable 'line-move-ignore-invisible)
--- 3454,3462 ----
    (make-local-variable 'Info-history)
    (make-local-variable 'Info-history-forward)
    (make-local-variable 'Info-index-alternatives)
!   (if Info-use-header-line    ; do not override global header lines
!       (setq header-line-format
! 	    '(:eval (get-text-property (point-min) 'header-line))))
    (set (make-local-variable 'tool-bar-map) info-tool-bar-map)
    ;; This is for the sake of the invisible text we use handling titles.
    (make-local-variable 'line-move-ignore-invisible)
