Index: src/s/darwin.h
===================================================================
RCS file: /cvsroot/emacs/emacs/src/s/darwin.h,v
retrieving revision 1.26
diff -c -r1.26 darwin.h
*** src/s/darwin.h	6 Feb 2006 15:23:23 -0000	1.26
--- src/s/darwin.h	10 Mar 2006 18:22:22 -0000
***************
*** 277,283 ****
     end of the header for adding load commands.  Needed for dumping.
     0x690 is the total size of 30 segment load commands (at 56
     each).  */
! #define LD_SWITCH_SYSTEM_TEMACS -prebind LIBS_CARBON -Xlinker -headerpad -Xlinker 690
  
  #define C_SWITCH_SYSTEM_TEMACS -Dtemacs
  
--- 277,283 ----
     end of the header for adding load commands.  Needed for dumping.
     0x690 is the total size of 30 segment load commands (at 56
     each).  */
! #define LD_SWITCH_SYSTEM_TEMACS -prebind LIBS_CARBON -Xlinker -headerpad -Xlinker 990
  
  #define C_SWITCH_SYSTEM_TEMACS -Dtemacs
  
