*** lisp/longlines.el	26 Feb 2007 20:23:04 +0000	1.33
--- lisp/longlines.el	10 May 2007 14:36:06 +0100	
***************
*** 412,418 ****
    "Re-wrap the buffer if the window width has changed.
  This is called by `window-configuration-change-hook'."
    (when (/= fill-column (- (window-width) window-min-width))
!     (setq fill-column (- (window-width) window-min-width))
      (let ((mod (buffer-modified-p)))
        (longlines-wrap-region (point-min) (point-max))
        (set-buffer-modified-p mod))))
--- 412,421 ----
    "Re-wrap the buffer if the window width has changed.
  This is called by `window-configuration-change-hook'."
    (when (/= fill-column (- (window-width) window-min-width))
!     ;; scale the fill column in order to minimize cases were 
!     ;; a wrap occurs because the average character width in a
!     ;; line is higher than expected.
!     (setq fill-column (floor (* 0.97 (- (window-width) window-min-width))))
      (let ((mod (buffer-modified-p)))
        (longlines-wrap-region (point-min) (point-max))
        (set-buffer-modified-p mod))))
