Index: mac-win.el
===================================================================
RCS file: /cvsroot/emacs/emacs/lisp/term/mac-win.el,v
retrieving revision 1.50
diff -c -r1.50 mac-win.el
*** mac-win.el	13 Jul 2005 09:11:35 -0000	1.50
--- mac-win.el	17 Jul 2005 20:42:22 -0000
***************
*** 1723,1728 ****
--- 1723,1736 ----
  ascii:-*-Monaco-*-*-*-*-12-*-*-*-*-*-mac-roman")
  (fontset-add-mac-fonts "fontset-mac" t)
  
+ (create-fontset-from-mac-roman-font 
+  "-apple-lucida grande*-medium-r-*-*-13-*-*-*-*-*-mac-roman"
+  nil
+  "lucida")
+  
+ (setq initial-frame-alist (cons (cons 'font "fontset-lucida") initial-frame-alist))
+ (fontset-add-mac-fonts "fontset-mac" nil)  
+   
  ;; Create fontset specified in X resources "Fontset-N" (N is 0, 1, ...).
  (create-fontset-from-x-resource)
  
***************
*** 1835,1840 ****
--- 1843,1849 ----
  				    (concat "file:" file-name))))
  	    (car (cdr (cdr event)))))
    (raise-frame))
+ 
  
  (global-set-key [drag-n-drop] 'mac-drag-n-drop)
  
