Index: lisp/menu-bar.el
===================================================================
RCS file: /sources/emacs/emacs/lisp/menu-bar.el,v
retrieving revision 1.300
diff -c -r1.300 menu-bar.el
*** lisp/menu-bar.el	10 Dec 2006 20:46:52 -0000	1.300
--- lisp/menu-bar.el	14 Dec 2006 16:41:08 -0000
***************
*** 74,79 ****
--- 74,97 ----
  (defvar uniquify-buffer-name-style)
  
  
+ 
+ (defun menu-bar-menu-frame-live-and-visible-p ()
+   "Return non-nil if the menu frame is alive and visible.
+ The menu frame is the frame for which we are updating the menu."
+   (let ((menu-frame (selected-frame)))
+     (and (frame-live-p menu-frame)
+ 	 ;; not icon
+ 	 (eq (frame-visible-p menu-frame) t))))
+ 
+ (defun menu-bar-non-minibuffer-window-p ()
+   "Return non-nil if selected window of the menu frame is not a minibuf window.
+ 
+ See the documentation of `menu-bar-menu-frame-live-and-visible-p'
+ for the definition of the menu frame."
+   (let ((menu-frame (or menu-updating-frame (selected-frame))))
+     (not (window-minibuffer-p (frame-selected-window menu-frame)))))
+ 
+ 
  ;; The "File" menu items
  (define-key menu-bar-file-menu [exit-emacs]
    '(menu-item "Exit Emacs" save-buffers-kill-emacs
***************
*** 172,177 ****
--- 190,196 ----
    '(menu-item "Save" save-buffer
  	      :enable (and (buffer-modified-p)
  			   (buffer-file-name)
+ 			   (menu-bar-menu-frame-live-and-visible-p)
  			   (menu-bar-non-minibuffer-window-p))
  	      :help "Save current buffer to its file"))
  
***************
*** 195,205 ****
  	      :help "Discard (kill) current buffer"))
  (define-key menu-bar-file-menu [insert-file]
    '(menu-item "Insert File..." insert-file
! 	      :enable (menu-bar-non-minibuffer-window-p)
  	      :help "Insert another file into current buffer"))
  (define-key menu-bar-file-menu [dired]
    '(menu-item "Open Directory..." dired
! 	      :enable (menu-bar-non-minibuffer-window-p)
  	      :help "Read a directory, to operate on its files"))
  (define-key menu-bar-file-menu [open-file]
    '(menu-item "Open File..." menu-find-file-existing
--- 214,226 ----
  	      :help "Discard (kill) current buffer"))
  (define-key menu-bar-file-menu [insert-file]
    '(menu-item "Insert File..." insert-file
! 	      :enable (and (menu-bar-non-minibuffer-window-p)
! 			   (menu-bar-menu-frame-live-and-visible-p))
  	      :help "Insert another file into current buffer"))
  (define-key menu-bar-file-menu [dired]
    '(menu-item "Open Directory..." dired
! 	      :enable (and (menu-bar-non-minibuffer-window-p)
! 			   (menu-bar-menu-frame-live-and-visible-p))
  	      :help "Read a directory, to operate on its files"))
  (define-key menu-bar-file-menu [open-file]
    '(menu-item "Open File..." menu-find-file-existing
***************
*** 368,374 ****
  
  ;;; Assemble the top-level Edit menu items.
  (define-key menu-bar-edit-menu [props]
!   '(menu-item "Text Properties" facemenu-menu))
  
  (define-key menu-bar-edit-menu [fill]
    '(menu-item "Fill" fill-region
--- 389,396 ----
  
  ;;; Assemble the top-level Edit menu items.
  (define-key menu-bar-edit-menu [props]
!   '(menu-item "Text Properties" facemenu-menu
! 	      :enable (menu-bar-menu-frame-live-and-visible-p)))
  
  (define-key menu-bar-edit-menu [fill]
    '(menu-item "Fill" fill-region
***************
*** 439,457 ****
  	      :help "Read a line number and go to that line"))
  
  (define-key menu-bar-edit-menu [goto]
!   (list 'menu-item "Go To" menu-bar-goto-menu))
  
  (define-key menu-bar-edit-menu [replace]
!   (list 'menu-item "Replace" menu-bar-replace-menu))
  
  (define-key menu-bar-edit-menu [search]
!   (list 'menu-item "Search" menu-bar-search-menu))
  
  (define-key menu-bar-edit-menu [separator-search]
    '(menu-item "--"))
  
  (define-key menu-bar-edit-menu [mark-whole-buffer]
    '(menu-item "Select All" mark-whole-buffer
  	      :help "Mark the whole buffer for a subsequent cut/copy."))
  (define-key menu-bar-edit-menu [clear]
    '(menu-item "Clear" delete-region
--- 461,483 ----
  	      :help "Read a line number and go to that line"))
  
  (define-key menu-bar-edit-menu [goto]
!   `(menu-item "Go To" ,menu-bar-goto-menu
! 	        :enable (menu-bar-menu-frame-live-and-visible-p)))
  
  (define-key menu-bar-edit-menu [replace]
!   `(menu-item "Replace" ,menu-bar-replace-menu
! 	      :enable (menu-bar-menu-frame-live-and-visible-p)))
  
  (define-key menu-bar-edit-menu [search]
!   `(menu-item "Search" ,menu-bar-search-menu
! 	      :enable (menu-bar-menu-frame-live-and-visible-p)))
  
  (define-key menu-bar-edit-menu [separator-search]
    '(menu-item "--"))
  
  (define-key menu-bar-edit-menu [mark-whole-buffer]
    '(menu-item "Select All" mark-whole-buffer
+ 	      :enable (menu-bar-menu-frame-live-and-visible-p)
  	      :help "Mark the whole buffer for a subsequent cut/copy."))
  (define-key menu-bar-edit-menu [clear]
    '(menu-item "Clear" delete-region
***************
*** 464,470 ****
  (fset 'yank-menu (cons 'keymap yank-menu))
  (define-key menu-bar-edit-menu [paste-from-menu]
    '(menu-item "Paste from kill menu" yank-menu
! 	      :enable (and (cdr yank-menu) (not buffer-read-only))
  	      :help "Choose a string from the kill ring and paste it"))
  (define-key menu-bar-edit-menu [paste]
    '(menu-item "Paste" yank
--- 490,497 ----
  (fset 'yank-menu (cons 'keymap yank-menu))
  (define-key menu-bar-edit-menu [paste-from-menu]
    '(menu-item "Paste from kill menu" yank-menu
! 	      :enable (and (cdr yank-menu) (not buffer-read-only)
! 			   (menu-bar-menu-frame-live-and-visible-p))
  	      :help "Choose a string from the kill ring and paste it"))
  (define-key menu-bar-edit-menu [paste]
    '(menu-item "Paste" yank
***************
*** 472,478 ****
  		       ;; Emacs compiled --without-x doesn't have
  		       ;; x-selection-exists-p.
  		       (fboundp 'x-selection-exists-p)
! 		       (x-selection-exists-p) (not buffer-read-only))
  	      :help "Paste (yank) text most recently cut/copied"))
  (define-key menu-bar-edit-menu [copy]
    '(menu-item "Copy" menu-bar-kill-ring-save
--- 499,506 ----
  		       ;; Emacs compiled --without-x doesn't have
  		       ;; x-selection-exists-p.
  		       (fboundp 'x-selection-exists-p)
! 		       (x-selection-exists-p) (not buffer-read-only)
! 		       (menu-bar-menu-frame-live-and-visible-p))
  	      :help "Paste (yank) text most recently cut/copied"))
  (define-key menu-bar-edit-menu [copy]
    '(menu-item "Copy" menu-bar-kill-ring-save
***************
*** 484,492 ****
--- 512,522 ----
  	      :enable (and mark-active (not buffer-read-only))
  	      :help
  	      "Cut (kill) text in region between mark and current position"))
+ 			   (menu-bar-menu-frame-live-and-visible-p)
  (define-key menu-bar-edit-menu [undo]
    '(menu-item "Undo" undo
  	      :enable (and (not buffer-read-only)
+ 			   (menu-bar-menu-frame-live-and-visible-p)
  			   (not (eq t buffer-undo-list))
  			   (if (eq last-command 'undo)
  			       (listp pending-undo-list)
***************
*** 1409,1429 ****
    '(menu-item "Emacs Tutorial" help-with-tutorial
  	      :help "Learn how to use Emacs"))
  
- (defun menu-bar-menu-frame-live-and-visible-p ()
-   "Return non-nil if the menu frame is alive and visible.
- The menu frame is the frame for which we are updating the menu."
-   (let ((menu-frame (or menu-updating-frame (selected-frame))))
-     (and (frame-live-p menu-frame)
- 	 (frame-visible-p menu-frame))))
- 
- (defun menu-bar-non-minibuffer-window-p ()
-   "Return non-nil if selected window of the menu frame is not a minibuf window.
- 
- See the documentation of `menu-bar-menu-frame-live-and-visible-p'
- for the definition of the menu frame."
-   (let ((menu-frame (or menu-updating-frame (selected-frame))))
-     (not (window-minibuffer-p (frame-selected-window menu-frame)))))
- 
  (defun kill-this-buffer ()	; for the menu bar
    "Kill the current buffer."
    (interactive)
--- 1439,1444 ----
***************
*** 1432,1442 ****
--- 1447,1459 ----
  (defun kill-this-buffer-enabled-p ()
    (let ((count 0)
  	(buffers (buffer-list)))
+ 	 (menu-bar-menu-frame-live-and-visible-p)
      (while buffers
        (or (string-match "^ " (buffer-name (car buffers)))
  	  (setq count (1+ count)))
        (setq buffers (cdr buffers)))
      (and (menu-bar-non-minibuffer-window-p)
+ 	 (menu-bar-menu-frame-live-and-visible-p)
  	 (> count 1))))
  
  (put 'dired 'menu-enable '(menu-bar-non-minibuffer-window-p))
