Index: src/minibuf.c
===================================================================
RCS file: /cvsroot/emacs/emacs/src/minibuf.c,v
retrieving revision 1.291
diff -c -r1.291 minibuf.c
*** src/minibuf.c	26 Nov 2005 11:19:21 -0000	1.291
--- src/minibuf.c	27 Nov 2005 15:14:45 -0000
***************
*** 1747,1754 ****
      XSETFASTINT (histpos, 0);
  
    val = read_minibuf (NILP (require_match)
! 		      ? Vminibuffer_local_completion_map
! 		      : Vminibuffer_local_must_match_map,
  		      init, prompt, make_number (pos), 0,
  		      histvar, histpos, def, 0,
  		      !NILP (inherit_input_method), 0);
--- 1747,1758 ----
      XSETFASTINT (histpos, 0);
  
    val = read_minibuf (NILP (require_match)
! 		      ? (NILP (Vminibuffer_completing_file_name) 
! 			 ? Vminibuffer_local_completion_map
! 			 : Vminibuffer_local_filename_completion_map)
! 		      : (NILP (Vminibuffer_completing_file_name) 
! 			 ? Vminibuffer_local_must_match_map
! 			 : Vminibuffer_local_must_match_filename_map),
  		      init, prompt, make_number (pos), 0,
  		      histvar, histpos, def, 0,
  		      !NILP (inherit_input_method), 0);
***************
*** 2921,2930 ****
--- 2925,2940 ----
    initial_define_key (Vminibuffer_local_completion_map, '?',
  		      "minibuffer-completion-help");
  
+   initial_define_key (Vminibuffer_local_filename_completion_map, ' ',
+ 		      "self-insert-command");
+ 
    initial_define_key (Vminibuffer_local_must_match_map, Ctl ('m'),
  		      "minibuffer-complete-and-exit");
    initial_define_key (Vminibuffer_local_must_match_map, Ctl ('j'),
  		      "minibuffer-complete-and-exit");
+ 
+   initial_define_key (Vminibuffer_local_must_match_filename_map, ' ', 
+ 		      "self-insert-command");
  }
  
  /* arch-tag: 8f69b601-fba3-484c-a6dd-ceaee54a7a73
Index: src/keymap.c
===================================================================
RCS file: /cvsroot/emacs/emacs/src/keymap.c,v
retrieving revision 1.309
diff -c -r1.309 keymap.c
*** src/keymap.c	21 Nov 2005 23:32:12 -0000	1.309
--- src/keymap.c	27 Nov 2005 15:14:46 -0000
***************
*** 65,70 ****
--- 65,77 ----
  /* was MinibufLocalCompletionMap */
  Lisp_Object Vminibuffer_local_completion_map;
  
+ /* keymap used for minibuffers when doing completion in filenames */
+ Lisp_Object Vminibuffer_local_filename_completion_map;
+ 
+ /* keymap used for minibuffers when doing completion in filenames 
+    with require-match*/
+ Lisp_Object Vminibuffer_local_must_match_filename_map;
+ 
  /* keymap used for minibuffers when doing completion and require a match */
  /* was MinibufLocalMustMatchMap */
  Lisp_Object Vminibuffer_local_must_match_map;
***************
*** 3782,3792 ****
--- 3785,3810 ----
    Vminibuffer_local_completion_map = Fmake_sparse_keymap (Qnil);
    Fset_keymap_parent (Vminibuffer_local_completion_map, Vminibuffer_local_map);
  
+   DEFVAR_LISP ("minibuffer-local-filename-completion-map", 
+ 	       &Vminibuffer_local_filename_completion_map,
+ 	       doc: /* Local keymap for minibuffer input with completion for filenames.  */);
+   Vminibuffer_local_filename_completion_map = Fmake_sparse_keymap (Qnil);
+   Fset_keymap_parent (Vminibuffer_local_filename_completion_map, 
+ 		      Vminibuffer_local_completion_map);
+ 
+ 
    DEFVAR_LISP ("minibuffer-local-must-match-map", &Vminibuffer_local_must_match_map,
  	       doc: /* Local keymap for minibuffer input with completion, for exact match.  */);
    Vminibuffer_local_must_match_map = Fmake_sparse_keymap (Qnil);
    Fset_keymap_parent (Vminibuffer_local_must_match_map,
  		      Vminibuffer_local_completion_map);
+ 
+   DEFVAR_LISP ("minibuffer-local-must-match-filename-map", 
+ 	       &Vminibuffer_local_must_match_filename_map,
+ 	       doc: /* Local keymap for minibuffer input with completion for filenames with exact match.  */);
+   Vminibuffer_local_must_match_filename_map = Fmake_sparse_keymap (Qnil);
+   Fset_keymap_parent (Vminibuffer_local_must_match_filename_map, 
+ 		      Vminibuffer_local_must_match_map);
  
    DEFVAR_LISP ("minor-mode-map-alist", &Vminor_mode_map_alist,
  	       doc: /* Alist of keymaps to use for minor modes.
Index: src/commands.h
===================================================================
RCS file: /cvsroot/emacs/emacs/src/commands.h,v
retrieving revision 1.22
diff -c -r1.22 commands.h
*** src/commands.h	7 Aug 2005 12:33:16 -0000	1.22
--- src/commands.h	27 Nov 2005 15:14:46 -0000
***************
*** 37,44 ****
--- 37,51 ----
  /* keymap used for minibuffers when doing completion */
  extern Lisp_Object Vminibuffer_local_completion_map;
  
+ /* keymap used for minibuffers when doing completion in filenames*/
+ extern Lisp_Object Vminibuffer_local_filename_completion_map;
+ 
  /* keymap used for minibuffers when doing completion and require a match */
  extern Lisp_Object Vminibuffer_local_must_match_map;
+ 
+ /* keymap used for minibuffers when doing completion in filenames
+    and require a match */
+ extern Lisp_Object Vminibuffer_local_must_match_filename_map;
  
  /* Last character of last key sequence.  */
  extern Lisp_Object last_command_char;
