*** ../emacs/lisp/term/mac-win.el	Tue Aug  2 20:02:14 2005
--- lisp/term/mac-win.el	Wed Aug 31 08:57:15 2005
***************
*** 1297,1303 ****
  (when (eq system-type 'darwin)
    (put 'FIND 'mac-scrap-name "com.apple.scrap.find")
    (put 'PRIMARY 'mac-scrap-name
!        (format "org.gnu.Emacs.%d.selection.PRIMARY" (emacs-pid))))
  (put 'com.apple.traditional-mac-plain-text 'mac-ostype "TEXT")
  (put 'public.utf16-plain-text 'mac-ostype "utxt")
  (put 'public.tiff 'mac-ostype "TIFF")
--- 1297,1303 ----
  (when (eq system-type 'darwin)
    (put 'FIND 'mac-scrap-name "com.apple.scrap.find")
    (put 'PRIMARY 'mac-scrap-name
!        (format "org.gnu.AquamacsEmacs.%d.selection.PRIMARY" (emacs-pid))))
  (put 'com.apple.traditional-mac-plain-text 'mac-ostype "TEXT")
  (put 'public.utf16-plain-text 'mac-ostype "utxt")
  (put 'public.tiff 'mac-ostype "TIFF")
