Index: lisp/bindings.el
===================================================================
RCS file: /sources/emacs/emacs/lisp/bindings.el,v
retrieving revision 1.166
diff -c -r1.166 bindings.el
*** lisp/bindings.el	6 Aug 2006 00:32:33 -0000	1.166
--- lisp/bindings.el	20 Aug 2006 11:41:27 -0000
***************
*** 273,278 ****
--- 273,281 ----
      map) "\
  Keymap to display on minor modes.")
  
+ (defvar command-line-processed nil)
+ (defvar before-init-hook nil)
+ (eval-at-startup ;; because of window-system
  (let* ((help-echo
  	;; The multi-line message doesn't work terribly well on the
  	;; bottom mode line...  Better ideas?
***************
*** 280,290 ****
  	;; mouse-1: select window, mouse-2: delete others, mouse-3: delete,
  	;; drag-mouse-1: resize, C-mouse-2: split horizontally"
  	"mouse-1: select (drag to resize), mouse-2: delete others, mouse-3: delete this")
!        (dashes (propertize "--" 'help-echo help-echo))
         (standard-mode-line-format
  	(list
  	 "%e"
! 	 (propertize "-" 'help-echo help-echo)
  	 'mode-line-mule-info
  	 'mode-line-modified
  	 'mode-line-frame-identification
--- 283,295 ----
  	;; mouse-1: select window, mouse-2: delete others, mouse-3: delete,
  	;; drag-mouse-1: resize, C-mouse-2: split horizontally"
  	"mouse-1: select (drag to resize), mouse-2: delete others, mouse-3: delete this")
!        (lotsofdashes (if window-system (make-string 100 32) "%-"))
!        (dash (propertize (if window-system " " "-") 'help-echo help-echo))
!        (dashes (propertize (if window-system "  " "--") 'help-echo help-echo))
         (standard-mode-line-format
  	(list
  	 "%e"
! 	 (propertize dash 'help-echo help-echo)
  	 'mode-line-mule-info
  	 'mode-line-modified
  	 'mode-line-frame-identification
***************
*** 296,302 ****
  	 'mode-line-modes
  	 `(which-func-mode ("" which-func-format ,dashes))
  	 `(global-mode-string (,dashes global-mode-string))
! 	 (propertize "-%-" 'help-echo help-echo)))
         (standard-mode-line-modes
  	(list
  	 (propertize "%[(" 'help-echo help-echo)
--- 301,307 ----
  	 'mode-line-modes
  	 `(which-func-mode ("" which-func-format ,dashes))
  	 `(global-mode-string (,dashes global-mode-string))
! 	 (propertize (concat dash lotsofdashes) 'help-echo help-echo)))
         (standard-mode-line-modes
  	(list
  	 (propertize "%[(" 'help-echo help-echo)
***************
*** 313,319 ****
  		     'mouse-face 'mode-line-highlight
  		     'local-map (make-mode-line-mouse-map
  				 'mouse-2 #'mode-line-widen))
! 	 (propertize ")%]--" 'help-echo help-echo)))
         (standard-mode-line-position
  	`((-3 ,(propertize "%p" 'help-echo help-echo))
  	  (size-indication-mode
--- 318,324 ----
  		     'mouse-face 'mode-line-highlight
  		     'local-map (make-mode-line-mouse-map
  				 'mouse-2 #'mode-line-widen))
! 	 (propertize (concat ")%]" dashes) 'help-echo help-echo)))
         (standard-mode-line-position
  	`((-3 ,(propertize "%p" 'help-echo help-echo))
  	  (size-indication-mode
***************
*** 335,341 ****
  
    (setq-default mode-line-position standard-mode-line-position)
    (put 'mode-line-position 'standard-value
!        (list `(quote ,standard-mode-line-position))))
  
  (defvar mode-line-buffer-identification-keymap nil "\
  Keymap for what is displayed by `mode-line-buffer-identification'.")
--- 340,346 ----
  
    (setq-default mode-line-position standard-mode-line-position)
    (put 'mode-line-position 'standard-value
!        (list `(quote ,standard-mode-line-position)))))
  
  (defvar mode-line-buffer-identification-keymap nil "\
  Keymap for what is displayed by `mode-line-buffer-identification'.")
