*** lisp/bindings.el	28 Oct 2005 17:19:16 +0100	1.154
--- lisp/bindings.el	18 Nov 2005 08:56:41 +0000	
***************
*** 288,298 ****
  	;; mouse-1: select window, mouse-2: delete others, mouse-3: delete,
  	;; drag-mouse-1: resize, C-mouse-2: split horizontally"
  	"mouse-1: select (drag to resize), mouse-2: delete others, mouse-3: delete this")
!        (dashes (propertize "--" 'help-echo help-echo)))
    (setq-default mode-line-format
      (list
       "%e"
!      (propertize "-" 'help-echo help-echo)
       'mode-line-mule-info
       'mode-line-modified
       'mode-line-frame-identification
--- 288,301 ----
  	;; mouse-1: select window, mouse-2: delete others, mouse-3: delete,
  	;; drag-mouse-1: resize, C-mouse-2: split horizontally"
  	"mouse-1: select (drag to resize), mouse-2: delete others, mouse-3: delete this")
!        (lotsofdashes (if window-system " " "%-"))
!        (dash (propertize (if window-system " " "-") 'help-echo help-echo))
!        (dashes (propertize (if window-system "  " "--") 'help-echo help-echo)))
! 
    (setq-default mode-line-format
      (list
       "%e"
!      (propertize dash 'help-echo help-echo)
       'mode-line-mule-info
       'mode-line-modified
       'mode-line-frame-identification
***************
*** 304,310 ****
       'mode-line-modes
       `(which-func-mode ("" which-func-format ,dashes))
       `(global-mode-string (,dashes global-mode-string))
!      (propertize "-%-" 'help-echo help-echo)))
  
    (setq-default mode-line-modes
      (list
--- 307,313 ----
       'mode-line-modes
       `(which-func-mode ("" which-func-format ,dashes))
       `(global-mode-string (,dashes global-mode-string))
!      (propertize (concat dash lotsofdashes) 'help-echo help-echo)))
  
    (setq-default mode-line-modes
      (list
***************
*** 322,328 ****
  		 'mouse-face 'mode-line-highlight
  		 'local-map (make-mode-line-mouse-map
  			     'mouse-2 #'mode-line-widen))
!      (propertize ")%]--" 'help-echo help-echo)))
  
    (setq-default mode-line-position
      `((-3 ,(propertize "%p" 'help-echo help-echo))
--- 325,331 ----
  		 'mouse-face 'mode-line-highlight
  		 'local-map (make-mode-line-mouse-map
  			     'mouse-2 #'mode-line-widen))
!      (propertize (concat ")%]" dashes) 'help-echo help-echo)))
  
    (setq-default mode-line-position
      `((-3 ,(propertize "%p" 'help-echo help-echo))
