Index: src/puresize.h
===================================================================
RCS file: /cvsroot/emacs/emacs/src/puresize.h,v
retrieving revision 1.74
diff -c -r1.74 puresize.h
*** src/puresize.h	7 Aug 2005 12:33:17 -0000	1.74
--- src/puresize.h	19 Oct 2005 23:34:05 -0000
***************
*** 37,42 ****
--- 37,47 ----
  #define SYSTEM_PURESIZE_EXTRA 0
  #endif
  
+ /* #define SITELOAD_PURESIZE_EXTRA 2000000 
+ */
+ 
+ #define SITELOAD_PURESIZE_EXTRA 700000
+ 
  #ifndef SITELOAD_PURESIZE_EXTRA
  #define SITELOAD_PURESIZE_EXTRA 0
  #endif
