*** lisp/textmodes/paragraphs.el	17 Apr 2008 10:52:44 +0100	1.87.2.4
--- lisp/textmodes/paragraphs.el	13 Jun 2008 15:04:21 +0100	
***************
*** 130,135 ****
--- 130,148 ----
    :group 'fill)
  ;;;###autoload(put 'sentence-end-double-space 'safe-local-variable 'booleanp)
  
+ (defcustom sentence-end-double-space-for-recognition nil
+   "Non-nil means a single space does not end a sentence.
+ This is relevant for the recognition of sentence ends.  See also
+ `sentence-end-without-period' and `colon-double-space'.  If non-nil, 
+ the value of `sentence-end-double-space' is used.
+ 
+ This value is used by the function `sentence-end' to construct the
+ regexp describing the end of a sentence, when the value of the variable
+ `sentence-end' is nil.  See Info node `(elisp)Standard Regexps'."
+   :type 'boolean
+   :group 'fill)
+ ;;;###autoload(put 'sentence-end-double-space-for-recognition 'safe-local-variable 'booleanp)
+ 
  (defcustom sentence-end-without-period nil
    "Non-nil means a sentence will end without a period.
  For example, a sentence in Thai text ends with double space but
***************
*** 188,194 ****
        (concat (if sentence-end-without-period "\\w  \\|")
  	      "\\("
  	      sentence-end-base
!               (if sentence-end-double-space
                    "\\($\\| $\\|\t\\|  \\)" "\\($\\|[\t ]\\)")
                "\\|[" sentence-end-without-space "]+"
  	      "\\)"
--- 201,208 ----
        (concat (if sentence-end-without-period "\\w  \\|")
  	      "\\("
  	      sentence-end-base
!               (if (and sentence-end-double-space
! 		       sentence-end-double-space-for-recognition)
                    "\\($\\| $\\|\t\\|  \\)" "\\($\\|[\t ]\\)")
                "\\|[" sentence-end-without-space "]+"
  	      "\\)"
