*** saveplace.el	22 Feb 2006 19:15:26 +0000	1.35
--- lisp/saveplace.el	07 Mar 2006 22:42:21 +0000	
***************
*** 207,213 ****
  (defun save-place-alist-to-file ()
    (let ((file (expand-file-name save-place-file)))
      (save-excursion
!       (message "Saving places to %s..." file)
        (set-buffer (get-buffer-create " *Saved Places*"))
        (delete-region (point-min) (point-max))
        (when save-place-forget-unreadable-files
--- 207,213 ----
  (defun save-place-alist-to-file ()
    (let ((file (expand-file-name save-place-file)))
      (save-excursion
!       ;; (message "Saving places to %s..." file)
        (set-buffer (get-buffer-create " *Saved Places*"))
        (delete-region (point-min) (point-max))
        (when save-place-forget-unreadable-files
***************
*** 225,233 ****
  	(condition-case nil
  	    ;; Don't use write-file; we don't want this buffer to visit it.
  	    (write-region (point-min) (point-max) file)
! 	  (file-error (message "Can't write %s" file)))
          (kill-buffer (current-buffer))
!         (message "Saving places to %s...done" file)))))
  
  (defun load-save-place-alist-from-file ()
    (if (not save-place-loaded)
--- 225,234 ----
  	(condition-case nil
  	    ;; Don't use write-file; we don't want this buffer to visit it.
  	    (write-region (point-min) (point-max) file)
! 	  (file-error (message "Saving places: Can't write %s" file)))
          (kill-buffer (current-buffer))
!         ;;(message "Saving places to %s...done" file)
! 	))))
  
  (defun load-save-place-alist-from-file ()
    (if (not save-place-loaded)
