*** lisp/startup.el	29 Feb 2008 05:10:09 +0000	1.436.2.14
--- lisp/startup.el	31 Mar 2008 23:01:43 +0100	
***************
*** 888,893 ****
--- 884,891 ----
  		      ;; into user-init-file.
  		      (setq user-init-file t)
  		      (load user-init-file-1 t t)
+ 		      (if (fboundp 'aquamacs-load-preferences) 
+ 			  (aquamacs-load-preferences))
  
  		      (when (eq user-init-file t)
  			;; If we did not find ~/.emacs, try
