Index: lisp/startup.el
===================================================================
RCS file: /sources/emacs/emacs/lisp/startup.el,v
retrieving revision 1.436.2.7
diff -c -r1.436.2.7 startup.el
*** lisp/startup.el	11 Nov 2007 20:23:23 -0000	1.436.2.7
--- lisp/startup.el	18 Dec 2007 01:59:15 -0000
***************
*** 889,894 ****
--- 889,896 ----
  		      ;; into user-init-file.
  		      (setq user-init-file t)
  		      (load user-init-file-1 t t)
+ 		      (if (fboundp 'aquamacs-load-preferences) 
+ 			  (aquamacs-load-preferences))
  
  		      (when (eq user-init-file t)
  			;; If we did not find ~/.emacs, try
***************
*** 1849,1881 ****
    (let ((resize-mini-windows t))
      (or noninteractive ;(input-pending-p) init-file-had-error
  	;; t if the init file says to inhibit the echo area startup message.
! 	(and inhibit-startup-echo-area-message
! 	     user-init-file
! 	     (or (and (get 'inhibit-startup-echo-area-message 'saved-value)
! 		      (equal inhibit-startup-echo-area-message
! 			     (if (equal init-file-user "")
! 				 (user-login-name)
! 			       init-file-user)))
! 		 ;; Wasn't set with custom; see if .emacs has a setq.
! 		 (let ((buffer (get-buffer-create " *temp*")))
! 		   (prog1
! 		       (condition-case nil
! 			   (save-excursion
! 			     (set-buffer buffer)
! 			     (insert-file-contents user-init-file)
! 			     (re-search-forward
! 			      (concat
! 			       "([ \t\n]*setq[ \t\n]+"
! 			       "inhibit-startup-echo-area-message[ \t\n]+"
! 			       (regexp-quote
! 				(prin1-to-string
! 				 (if (equal init-file-user "")
! 				     (user-login-name)
! 				   init-file-user)))
! 			       "[ \t\n]*)")
! 			      nil t))
! 			 (error nil))
! 		     (kill-buffer buffer)))))
  	(message "%s" (startup-echo-area-message)))))
  
  (defun display-startup-screen (&optional concise)
--- 1851,1857 ----
    (let ((resize-mini-windows t))
      (or noninteractive ;(input-pending-p) init-file-had-error
  	;; t if the init file says to inhibit the echo area startup message.
! 	inhibit-startup-echo-area-message
  	(message "%s" (startup-echo-area-message)))))
  
  (defun display-startup-screen (&optional concise)
