Index: src/macterm.c
===================================================================
RCS file: /cvsroot/emacs/emacs/src/macterm.c,v
retrieving revision 1.214.2.12
diff -c -p -r1.214.2.12 macterm.c
*** src/macterm.c       8 Jan 2008 04:29:56 -0000       1.214.2.12
--- src/macterm.c       14 Jan 2008 03:03:41 -0000
*************** void
*** 5581,5588 ****
  x_scroll_bar_clear (f)
       FRAME_PTR f;
  {
!   XTcondemn_scroll_bars (f);
!   XTjudge_scroll_bars (f);
  }
  
  
--- 5581,5600 ----
  x_scroll_bar_clear (f)
       FRAME_PTR f;
  {
!   Lisp_Object bar;
! 
!   /* We can have scroll bars even if this is 0,
!      if we just turned off scroll bar mode.
!      But in that case we should not clear them.  */
!   if (FRAME_HAS_VERTICAL_SCROLL_BARS (f))
!     {
! #if USE_CG_DRAWING
!       mac_prepare_for_quickdraw (f);
! #endif
!       for (bar = FRAME_SCROLL_BARS (f); VECTORP (bar);
!          bar = XSCROLL_BAR (bar)->next)
!       Draw1Control (SCROLL_BAR_CONTROL_HANDLE (XSCROLL_BAR (bar)));
!     }
  }
  
  