--- src/macterm.c	2008-07-11 15:16:15.000000000 -0400
+++ src/macterm.c	2008-07-11 15:11:48.000000000 -0400
@@ -5763,12 +5763,6 @@

  check_frame_size (f, &rows, &cols);

-  /* Refuse to set cols, rows, or both if in full-screen mode. */
-  if (f->want_fullscreen & FULLSCREEN_HEIGHT)
-    rows = FRAME_LINES (f);
-  if (f->want_fullscreen & FULLSCREEN_WIDTH)
-    cols = FRAME_COLS (f);
-
  f->scroll_bar_actual_width
    = FRAME_SCROLL_BAR_COLS (f) * FRAME_COLUMN_WIDTH (f);

@@ -5780,2 +5774,2 @@
  f->win_gravity = NorthWestGravity;
  x_wm_set_size_hint (f, (long) 0, 0);

+  /* Refuse to change height, width, or both if in full-screen mode. */
+  Rect b;
+  OSStatus st = GetWindowBounds(FRAME_MAC_WINDOW (f), kWindowContentRgn, &b);
+  if (st == noErr)
+    {
+      if (f->want_fullscreen & FULLSCREEN_HEIGHT)
+        pixelheight = b.bottom - b.top;
+      if (f->want_fullscreen & FULLSCREEN_WIDTH)
+        pixelwidth = b.right - b.left;
+    }
+
  mac_size_window (FRAME_MAC_WINDOW (f), pixelwidth, pixelheight, 0);

#if TARGET_API_MAC_CARBON
Index: src/mactoolbox.c
===================================================================
RCS file: /cvsroot/emacs/emacs/src/mactoolbox.c,v
retrieving revision 1.1.2.4
diff -c -p -r1.1.2.4 mactoolbox.c
*** src/mactoolbox.c	23 Apr 2008 08:56:20 -0000	1.1.2.4
--- src/mactoolbox.c	4 Jul 2008 03:20:27 -0000
*************** mac_handle_text_input_event (next_handle
*** 575,581 ****
 	    int hpos, vpos, x, y;
 	    struct glyph_row *row;
 	    struct glyph *glyph;
! 	    XFontStruct *font;

 	    f = mac_focus_frame (&one_mac_display_info);
 	    w = XWINDOW (f->selected_window);
--- 575,581 ----
 	    int hpos, vpos, x, y;
 	    struct glyph_row *row;
 	    struct glyph *glyph;
! 	    struct face *face;

 	    f = mac_focus_frame (&one_mac_display_info);
 	    w = XWINDOW (f->selected_window);
*************** mac_handle_text_input_event (next_handle
*** 600,608 ****
 		   + row->visible_height
 		   + f->top_pos + FRAME_OUTER_TO_INNER_DIFF_Y (f));

! 	    font = FACE_FROM_ID (f, glyph->face_id)->font;
! 	    if (font)
 	      {
 		Fixed point_size = Long2Fix (font->mac_fontsize);
 		short height = row->visible_height;
 		short ascent = row->ascent;
--- 600,609 ----
 		   + row->visible_height
 		   + f->top_pos + FRAME_OUTER_TO_INNER_DIFF_Y (f));

! 	    face = FACE_FROM_ID (f, glyph->face_id);
! 	    if (face && face->font)
 	      {
+ 		XFontStruct *font = face->font;
 		Fixed point_size = Long2Fix (font->mac_fontsize);
 		short height = row->visible_height;
 		short ascent = row->ascent;
