*** lisp/recentf.el	06 Apr 2008 13:51:37 +0200	1.56.2.2
--- lisp/recentf.el	10 Aug 2008 19:04:41 +0200	
***************
*** 394,400 ****
    "Convert file NAME to absolute, and canonicalize it.
  NAME is first passed to the function `expand-file-name', then to
  `recentf-filename-handlers' to post process it."
!   (recentf-apply-filename-handlers (expand-file-name name)))
  
  (defun recentf-include-p (filename)
    "Return non-nil if FILENAME should be included in the recent list.
--- 394,402 ----
    "Convert file NAME to absolute, and canonicalize it.
  NAME is first passed to the function `expand-file-name', then to
  `recentf-filename-handlers' to post process it."
!   (condition-case nil
!       (recentf-apply-filename-handlers (expand-file-name name))
!     (error name)))
  
  (defun recentf-include-p (filename)
    "Return non-nil if FILENAME should be included in the recent list.
*** lisp/erc/erc.el	28 Jul 2008 23:15:49 +0100	1.52
--- lisp/erc/erc.el	17 Aug 2008 19:44:33 +0100	
***************
*** 6085,6092 ****
  	(let ((header (if erc-header-line-format
  			  (format-spec erc-header-line-format spec)
  			nil)))
! 	  (cond ((null header)
! 		 (setq header-line-format nil))
  		(erc-header-line-uses-help-echo-p
  		 (let ((help-echo (with-temp-buffer
  				    (insert header)
--- 6085,6091 ----
  	(let ((header (if erc-header-line-format
  			  (format-spec erc-header-line-format spec)
  			nil)))
! 	  (cond ((null header))
  		(erc-header-line-uses-help-echo-p
  		 (let ((help-echo (with-temp-buffer
  				    (insert header)
