*** src/macterm.c.~1.210.~	Sat Mar 10 15:26:33 2007
--- src/macterm.c	Wed Mar 14 19:55:12 2007
*************** note_mouse_movement (frame, pos)
*** 4501,4507 ****
  	    rif->define_frame_cursor (frame,
  				      frame->output_data.mac->nontext_cursor);
  	}
-       return 1;
      }
    /* Has the mouse moved off the glyph it was on at the last sighting?  */
    if (frame != last_mouse_glyph_frame
--- 4501,4506 ----
