*** lisp/isearch.el	06 Apr 2008 12:51:37 +0100	1.297.2.8
--- lisp/isearch.el	13 May 2008 19:17:51 +0100	
***************
*** 1381,1386 ****
--- 1381,1389 ----
  	    ;; If the string was found but was completely invisible,
  	    ;; it might now be partly visible, so try again.
  	    (prog1 isearch-hidden (setq isearch-hidden nil)))
+     (if (and transient-mark-mode 
+ 	     (not (and cua-mode cua--explicit-region-start)))
+ 	(deactivate-mark))
      ;; In reverse search, adding stuff at
      ;; the end may cause zero or many more chars to be
      ;; matched, in the string following point.
***************
*** 1411,1418 ****
  		       (min isearch-opoint
  			    isearch-barrier
  			    (1+ isearch-other-end)))))
!       (isearch-search)
!       ))
    (isearch-push-state)
    (if isearch-op-fun (funcall isearch-op-fun))
    (isearch-update))
--- 1414,1420 ----
  		       (min isearch-opoint
  			    isearch-barrier
  			    (1+ isearch-other-end)))))
!       (isearch-search)))
    (isearch-push-state)
    (if isearch-op-fun (funcall isearch-op-fun))
    (isearch-update))
*** lisp/menu-bar.el	06 Apr 2008 12:51:37 +0100	1.302.2.11
--- lisp/menu-bar.el	14 May 2008 19:16:47 +0100	
***************
*** 519,532 ****
  (defun clipboard-kill-ring-save (beg end)
    "Copy region to kill ring, and save in the X clipboard."
    (interactive "r")
!   (let ((x-select-enable-clipboard t))
!     (kill-ring-save beg end)))
  
  (defun clipboard-kill-region (beg end)
    "Kill the region, and save it in the X clipboard."
    (interactive "r")
!   (let ((x-select-enable-clipboard t))
!     (kill-region beg end)))
  
  (defun menu-bar-enable-clipboard ()
    "Make CUT, PASTE and COPY (keys and menu bar items) use the clipboard.
--- 549,564 ----
  (defun clipboard-kill-ring-save (beg end)
    "Copy region to kill ring, and save in the X clipboard."
    (interactive "r")
!   (when (or (not transient-mark-mode) mark-active)
!     (let ((x-select-enable-clipboard t))
!       (kill-ring-save beg end))))
  
  (defun clipboard-kill-region (beg end)
    "Kill the region, and save it in the X clipboard."
    (interactive "r")
!   (when (or (not transient-mark-mode) mark-active)
!     (let ((x-select-enable-clipboard t))
!       (kill-region beg end))))
  
  (defun menu-bar-enable-clipboard ()
    "Make CUT, PASTE and COPY (keys and menu bar items) use the clipboard.
*** lisp/mouse.el	06 Apr 2008 12:51:37 +0100	1.315.2.7
--- lisp/mouse.el	14 May 2008 19:51:54 +0100	
***************
*** 949,955 ****
      (mouse-move-drag-overlay mouse-drag-overlay start-point start-point
                               click-count)
      (overlay-put mouse-drag-overlay 'window start-window)
!     (deactivate-mark)
      (let (event end end-point last-end-point)
        (track-mouse
  	(while (progn
--- 956,963 ----
      (mouse-move-drag-overlay mouse-drag-overlay start-point start-point
                               click-count)
      (overlay-put mouse-drag-overlay 'window start-window)
!     (unless (and cua-mode cua--explicit-region-start)
!       (deactivate-mark))
      (let (event end end-point last-end-point)
        (track-mouse
  	(while (progn
