diff -cr emacs.org/src/macterm.c emacs/src/macterm.c
*** emacs.org/src/macterm.c	Sat Apr 16 05:56:29 2005
--- emacs/src/macterm.c	Sat Apr 16 06:00:07 2005
***************
*** 304,309 ****
--- 304,314 ----
  extern void menubar_selection_callback (FRAME_PTR, int);
  extern void set_frame_menubar (FRAME_PTR, int, int);
  
+ #ifdef USE_TRANSPARENCY
+ /* Make Transparency */
+ int mac_transparency_alpha;
+ #endif
+ 
  /* X display function emulation */
  
  void
***************
*** 5242,5247 ****
--- 5247,5256 ----
    f->win_gravity = NorthWestGravity;
    x_wm_set_size_hint (f, (long) 0, 0);
  
+ #ifdef USE_TRANSPARENCY
+   if( 0<= mac_transparency_alpha && mac_transparency_alpha <= 100 )SetWindowAlpha (f->output_data.mac->mWP, ((float)mac_transparency_alpha/100));
+ #endif
+   
    SizeWindow (FRAME_MAC_WINDOW (f), pixelwidth, pixelheight, 0);
  #if TARGET_API_MAC_CARBON
    if (f->output_data.mac->hourglass_control)
***************
*** 9835,9840 ****
--- 9844,9856 ----
    Vmac_charset_info_alist =
      Fcons (list3 (build_string ("mac-roman"),
  		  make_number (smRoman), Qnil), Qnil);
+ 
+ #ifdef USE_TRANSPARENCY
+   DEFVAR_INT ("mac-transparency-alpha", &mac_transparency_alpha,
+     doc: /* Transparency ratio, can be set integer between 0 to 100.
+ 0 means fully transparent, 100 means non transparent */);
+   mac_transparency_alpha = 100;
+ #endif
  }
  
  /* arch-tag: f2259165-4454-4c04-a029-a133c8af7b5b
Only in emacs/src: macterm.c~
diff -cr emacs.org/src/w32term.c emacs/src/w32term.c
*** emacs.org/src/w32term.c	Sat Apr 16 05:56:43 2005
--- emacs/src/w32term.c	Sat Apr 16 05:51:09 2005
***************
*** 268,273 ****
--- 268,278 ----
  
  static Lisp_Object Qvendor_specific_keysyms;
  
+ #ifdef USE_TRANSPARENCY
+ /* Make Transparency */
+ int w32_transparency_alpha;
+ #endif
+ 
  
  /***********************************************************************
  			      Debugging
***************
*** 5537,5542 ****
--- 5542,5556 ----
      AdjustWindowRect(&rect, f->output_data.w32->dwStyle,
  		     FRAME_EXTERNAL_MENU_BAR (f));
  
+ #ifdef USE_TRANSPARENCY
+     SetWindowLong(FRAME_W32_WINDOW (f),GWL_EXSTYLE,WS_EX_LAYERED);
+     if( 0<= w32_transparency_alpha && w32_transparency_alpha <= 100 )
+       SetLayeredWindowAttributes(FRAME_W32_WINDOW (f),
+                                  RGB(255, 255, 255),
+                                  (int)((float)w32_transparency_alpha/(float)100*(float)255),
+                                  LWA_ALPHA);
+ #endif
+     
      my_set_window_pos (FRAME_W32_WINDOW (f),
  		       NULL,
  		       0, 0,
***************
*** 6530,6535 ****
--- 6543,6555 ----
  
  When this variable is set, other variables affecting the appearance of
  the cursor have no effect.  */);
+ 
+ #ifdef USE_TRANSPARENCY
+   DEFVAR_INT ("w32-transparency-alpha", &w32_transparency_alpha,
+              doc: /* Transparency ratio, can be set integer between 0 to 100.
+ 0 means fully transparent, 100 means non transparent */);
+   w32_transparency_alpha = 100;
+ #endif
  
    /* Initialize w32_use_visible_system_caret based on whether a screen
       reader is in use.  */
