*** /Users/dr/Projects/emacs/lisp/term/mac-win.el	Mon May 23 08:46:44 2005
--- ./mac-win.el	Mon Jun 13 19:57:12 2005
***************
*** 1660,1665 ****
--- 1660,1670 ----
  (create-fontset-from-fontset-spec
   "-etl-fixed-medium-r-normal-*-16-*-*-*-*-*-fontset-mac,
  ascii:-*-Monaco-*-*-*-*-12-*-*-*-*-*-mac-roman")
+ (create-fontset-from-fontset-spec 
+ "-etl-variable-medium-r-normal-*-*-*-*-*-*-*-fontset-lucida,
+ ascii:-*-lucida grande*-medium-r-*-*-*-*-*-*-*-*-mac-roman")
+ (setq initial-frame-alist (cons (cons 'font "fontset-lucida") initial-frame-alist))
+ 
  (fontset-add-mac-fonts "fontset-mac")
  
  ;; Create fontset specified in X resources "Fontset-N" (N is 0, 1, ...).
